---
title: "BKL_fungalDEGS"
author: "Britt"
date: "2024-07-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

To do:

heatmap with website from zack

pca from zach...know what the outlier is

pca from siyu

blast sequence from genes of interest (top 10 differentially expressed)

experiement with using smaller dataset (filter genes much more heavily), then redo DGE

can also use smaller dataset to experiment with heatmap

PERMANOVA betadisper (Laura)

watering regime and lysemitry time frame (in figure title)

units for heatmap?

format references: capitalize first letter of title, italicize species names

**Basic steps for DGE Analysis:**

1.  Load the RNAseq counts.
2.  Normalize the counts.
3.  QC the counts.
4.  Create a data frame that describes the experiment.
5.  Determine the dispersion parameter (how over-dispersed is the data?)
6.  Fit a statistical model for gene expression as a function of experimental parameters.
7.  Test the significance of experimental parameters for explaining gene expression.
8.  Examine result.

Start:

**1. Load the RNAseq counts.**

```{bash, eval = FALSE}
#take a look at fungal counts data
head TC_counts.tsv #replace this file with fungal counts that you get back from core
```

```{r}
library(readr)
fungal.counts <- read_tsv("TC_counts.tsv") #assign fungal counts data to an object
fungal.counts

fungal.counts.no30 <- read.csv("counts_data_no30 - counts_data.csv")
fungal.counts.no30
```

```{r}
#retain only the reads that were over 10 count in at least three samples
fungal.counts <- fungal.counts[rowSums(fungal.counts[,-1] > 10) >= 3,]

fungal.counts
```

```{r}
write.csv(fungal.counts, "filtered.fungal.counts.csv")
```

```{r}
#retain only the reads that were over 10 count in at least three samples
fungal.counts.no30 <- fungal.counts.no30[rowSums(fungal.counts.no30[,-1] > 10) >= 3,]

fungal.counts.no30
```

"We expect that read counts, especially from biological replicates, will be highly correlated. Check to see if this is the case using the `pairs()` function and the `cor()` function. Pairs should be applied to the (potentially log transformed) fungal.counts, NOT the correlation matrix."

**Important Hint:** *`pairs` is slow on the full dataset. Try it on the first 1,000 genes. Do you need to transform to make the pairs output more meaningful?*

**Important Hint2:** *it will be hard to see the pairs plot in the Rstudio inline display. Once you have the plot, click the expand to full window icon to display the plot in its own window.* \*\*Alternatively, instead of using all columns of data, try it on a smaller number of columns \*\*

**Hint 3:** *remember that you will need to remove the "gene_id" column before using the data in `pairs` or `cor`*

```{r}
fungal.counts.noid <- fungal.counts[, -which(names(fungal.counts) == "target_id")]
#remove target_id column

fungal.counts.noid

fungal.counts.noidno30 <- fungal.counts[, -which(names(fungal.counts.no30) == "target_id")]
#remove target_id column

fungal.counts.noidno30
```

```{r}
transformed.counts <- log2(fungal.counts.noid + 1) 
#log transform counts
```

```{r}
cor.table <- cor(fungal.counts.noid)
pairs(transformed.counts)
```

"Once you have a correlation table, use the code provided to visualize it. Then, comment on the results from the pairs and correlation heatmap plots. Are the replicates more similar to each other than they are to other samples? Do you think there are any mistakes in the sample treatment labels?"

```{r}
#rownames(cor.table) <- str_remove(rownames(cor.table), "_INTERNODE.*") #shorter names for better plotting
#colnames(cor.table) <- str_remove(colnames(cor.table), "_INTERNODE.*")

library(tidyverse)

cor.table %>% gplots::heatmap.2(dendrogram="row", trace = "none", col=viridis::viridis(25, begin=.25), margins=c(7,8))
```

**2. Normalize the counts and 3. QC the counts and 4. Make dataframe.**

We will use TTM normalization.

We need to tell EdgeR what groups our samples belong to. In order to do this we'll make a dataframe with our treatments and treatment combos. (drought treatment, inoculation treatment, and combination of both)

```{r}
#make tibble with descriptions from counts data table
sample.description <- tibble(sample=colnames(fungal.counts.noid))

sample.description
```

make three more columns:

1\. column "sp" that has TC or TCS, indicating the inoculation treatment. TC is Trunocolumella citrina only and TCS is T. citrina and Suillus both.

2\. column "trt" that indicates the treatment with either "C" or "T" (control or drought treatment)

3\. column "group" that has sp and trt pasted together with "\_" as a separator. You can use `str_c()` and the "sp" and "trt" columns for this.

```{r}
library(dplyr)
library(stringr)

sample.description <- sample.description %>%
  mutate(
    sp = case_when(
      str_detect(sample, "TCS") ~ "TCS",
      str_detect(sample, "TC") ~ "TC",
      TRUE ~ NA_character_
    ),
    trt = case_when(
      str_detect(sample, "_C") ~ "C",
      str_detect(sample, "_T") ~ "T",
      TRUE ~ NA_character_
    ),
    group = str_c(sp, trt, sep = "_")
  )

sample.description
```

```{r}
write.csv(sample.description, "sample.description.csv")
```

```{r}
sample.description <- sample.description %>%
  mutate(sp=factor(sp), 
         trt=factor(trt,levels = c("C","T"))) # setting the levels in this way makes "C" the reference  

sample.description
```

Calculate Normalization factors:

EdgeR wants numeric data, so we must convert it first. Then we calculate normalization factors using TTM method.

```{r}
#this part seems off
library(edgeR)
counts.matrix <- fungal.counts %>% select(-target_id) %>% as.matrix()
rownames(counts.matrix) <- fungal.counts$target_id

dge.data <- DGEList(counts=counts.matrix, 
                    group=sample.description$group)
dim(dge.data) 
dge.data <- calcNormFactors(dge.data, method = "TMM")
dge.data$samples # look at the normalization factors
```

We can now make an MDS plot to visualize how similar/different our samples our in gene expression. THIS is the bare minimum we should accomplish and put on the poster, but ideally we will dive further into analysis depending on how much time we have.

Make an MDS plot:

```{r}
plotMDS(dge.data)
```

Make a nicer MDS plot:

```{r}
mdsvals <- plotMDS(dge.data, plot = FALSE) # get the MDS values for plotting

tibble(x=mdsvals$x, y=mdsvals$y, sample=rownames(dge.data$samples)) %>%
  inner_join(sample.description) %>%
  ggplot(aes(x=x, y=y, color=trt, shape=sp)) +
  geom_point(size=3) 
```

Extract normalized data:

I'm still not sure why we would log transform the normalized data or not. Here is how we extract the normalized data either way.

```{r}
fungal.counts.normal <- cpm(dge.data) 

# or log2 transformed:
fungal.counts.normal.log <- cpm(dge.data,log = TRUE)
```

**5. Determine the dispersion parameter (how over-dispersed is the data?)**

Calculate dispersion factors:

First we tell EdgeR what our experimental design was. Then, the dispersions.

```{r}
#sample.description is the original descriptive dataset (describes our samples but doesn't have counts data)
design <- model.matrix(~sp+trt,data = sample.description)
rownames(design) <- sample.description$sample
design
```

```{r}
#First the overall dispersion
dge.data <- estimateGLMCommonDisp(dge.data,design,verbose = TRUE)

#Then a trended dispersion based on count level
dge.data <- estimateGLMTrendedDisp(dge.data,design)

#And lastly we calculate the gene-wise dispersion, using the prior estimates to "squeeze" the dispersion towards the common dispersion.
dge.data <- estimateGLMTagwiseDisp(dge.data,design)

#We can examine this with a plot
plotBCV(dge.data)
```

**6. Fit a statistical model for gene expression as a function of experimental parameters.**

To find differentially expressed genes, we basically make linear models for the data that have more or less coefficients. (We compare the full model, with all coefficients, with a model that has one or more coefficients dropped to look at the difference between them.)

```{r}
#this is the full model with all coefficients
#dge.data is the normalized count data
fit <- glmFit(dge.data, design)
```

```{r}
#this model looks at genes differentially expressed between two treatments
trt.lrt <- glmLRT(fit,coef = "trtT")
```

```{r}
#this model between two species
sp.lrt <- glmLRT(fit,coef = "spTCS")
```

**7. Test the significance of experimental parameters for explaining gene expression.**

```{r}
top10.trt <- topTags(trt.lrt) # the top 10 most differentially expressed genes between the two treatments (using the above model)
top10.trt
```

```{r}
top10.sp <- topTags(sp.lrt) # the top 10 most differentially expressed genes between the two species (using the above model)
top10.sp
```

```{r}
top10.sp.trt <- topTags(trt.lrt.interaction) # the top 10 most differentially expressed genes between the two species (using the above model)
top10.sp.trt
test=trt.lrt$table
FDR<- p.adjust(test$PValue,method="bonferroni") 
```

```{r}
write.csv(top10.sp.trt,"top10.sp.trt.csv")
write.csv(top10.sp,"top10.sp.csv")
write.csv(top10.trt, "top10.trt.csv")
```

In the resulting table from using topTags,

-   logFC is the log2 fold-change in expression between R500 and IMB211. So a logFC of 2 indicates that the gene is expressed 4 times higher in R500; a logFC of -3 indicates that it is 8 times lower in R500.

-   logCPM is the average expression across all samples.

-   LR is the likelihood ratio: L(Full model) / L(small model) .

-   PValue: unadjusted p-value

-   FDR: False discovery rate (p-value adjusted for multiple testing...this is the one to use)

To summarize:

```{r}
summary(decideTestsDGE(trt.lrt,p.value=0.99)) #This uses the FDR.  0.05 would be OK also.
```

```{r}
summary(decideTestsDGE(sp.lrt,p.value=0.01)) #This uses the FDR.  0.05 would be OK also.
```

In the table resulting from decideTestsDGE, the Down row is the number of down regulated genes in R500 relative to IMB211 and the Up row is the number of up regulated genes.

If we want to create a table of all differentially expressed genes at a given FDR, then:

```{r}
#Extract genes with a FDR < 0.01 (could also use 0.05)
DEgene.trt <- topTags(trt.lrt,n = Inf,p.value = 0.01)$table

#save to a file
write.csv(DEgene.trt,"DEgenes.trt.csv")

#Or if we want all genes, regardless of FDR:
DEgene.trt.all <- topTags(trt.lrt,n = Inf, p.value = 1)$table
#save to a file
write.csv(DEgene.trt.all,"DEgenes.trt.all.csv")
```

DID NOT GO FURTHER THAN BELOW THIS LINE

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

**8. Examine the result.**

Define a function to make a bar graph of genes of interest:

```{r}
plotDE <- function(genes, dge, sample.description) {
  require(ggplot2)
  tmp.data <- t(log2(cpm(dge[genes,])+1))
  tmp.data <- tmp.data %>%
    as.data.frame() %>%
    rownames_to_column("sample") %>%
    left_join(sample.description,by="sample")
  tmp.data <- tmp.data %>%
    pivot_longer(cols=starts_with("Bra"), values_to = "log2_cpm", names_to = "gene")
  pl <- ggplot(tmp.data,aes(x=gt,y=log2_cpm,fill=trt))
  pl <- pl + facet_wrap( ~ gene)
  pl <- pl + ylab("log2(cpm)") + xlab("genotype")
  pl <- pl + geom_boxplot()
  pl + theme(axis.text.x  = element_text(angle=45, vjust=1,hjust=1))
}
```

Use the function plotDE to plot a single gene of interest or the top 9 genes expressed etc:

```{r}
#change according to our genes of interest etc
# A single gene
plotDE("Bra009785",dge.data,sample.description)

#top 9 genes
plotDE(rownames(DEgene.gt)[1:9],dge.data,sample.description)
```

**Gene by Treatment Interaction**

Treatments have effects on treatments. One species of fungus may express genes in response to drought differently than a different species of fungus with the same drought treatment. We can find the genes that change in expression in this way by making a new model that includes this interaction of treatments.

Make the new model:

```{r}
design.interaction <- model.matrix(~sp*trt,data = sample.description)
rownames(design.interaction) <- sample.description$sample
design.interaction
```

Now using the new model, do the same steps that we did before to find differentially expressed genes:

```{r}
#First the overall dispersion
dge.data.interaction <- estimateGLMCommonDisp(dge.data,design.interaction,verbose = TRUE)

#Then a trended dispersion based on count level
dge.data.interaction <- estimateGLMTrendedDisp(dge.data,design.interaction)

#And lastly we calculate the gene-wise dispersion, using the prior estimates to "squeeze" the dispersion towards the common dispersion.
dge.data.interaction <- estimateGLMTagwiseDisp(dge.data,design.interaction)

#We can examine this with a plot
plotBCV(dge.data)
```

```{r}
fit.interaction <- glmFit(dge.data.interaction, design.interaction)
```

```{r}
#change coefficients as necessary
trt.lrt.interaction <- glmLRT(fit.interaction,coef = "spTCS:trtT")
```

```{r}
topTags(trt.lrt.interaction) # the top 10 most differentially expressed genes
```

```{r}
summary(decideTestsDGE(trt.lrt.interaction,p.value=0.01)) #This uses the FDR.  0.05 would be OK also.
```

```{r}
#Extract genes with a FDR < 0.01 (could also use 0.05)
DEgene.trt.interaction <- topTags(trt.lrt.interaction,n = Inf,p.value = 0.01)$table

#save to a file
write.csv(DEgene.trt.interaction,"../output/DEgenes.trt.interaction.csv")

#Or if we want all genes, regardless of FDR:
DEgene.trt.all.interaction <- topTags(trt.lrt.interaction,n = Inf, p.value = 1)$table
#save to a file
write.csv(DEgene.trt.all.interaction,"../output/DEgenes.trt.all.interaction.csv")
```

We already made the function "plotDE"...I need to double check that the below chunk of code doesn't need to be changed for an interaction model. It seems like it doesn't.

```{r}
plotDE <- function(genes, dge, sample.description) {
  require(ggplot2)
  tmp.data <- t(log2(cpm(dge[genes,])+1))
  tmp.data <- tmp.data %>%
    as.data.frame() %>%
    rownames_to_column("sample") %>%
    left_join(sample.description,by="sample")
  tmp.data <- tmp.data %>%
    pivot_longer(cols=starts_with("Bra"), values_to = "log2_cpm", names_to = "gene")
  pl <- ggplot(tmp.data,aes(x=gt,y=log2_cpm,fill=trt))
  pl <- pl + facet_wrap( ~ gene)
  pl <- pl + ylab("log2(cpm)") + xlab("genotype")
  pl <- pl + geom_boxplot()
  pl + theme(axis.text.x  = element_text(angle=45, vjust=1,hjust=1))
}
```

Use plotDE to plot gene expression just like we did for the non-interaction model:

```{r}
# A single gene
plotDE("Bra009785",dge.data.interaction,fungal.counts.new)

#top 9 genes
plotDE(rownames(DEgene.gt.interaction)[1:9],dge.data.interaction,fungal.counts.new)
```

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

Have NOT tried to do anything below this line yet.

Let's try with the biocore version instead...

## Differential Expression Analysis with Limma-Voom

**limma** is an R package that was originally developed for differential expression (DE) analysis of gene expression microarray data.

**voom** is a function in the limma package that transforms RNA-Seq data for use with limma.

Together they allow fast, flexible, and powerful analyses of RNA-Seq data. Limma-voom is *our* tool of choice for DE analyses because it:

-   Allows for incredibly flexible model specification (you can include multiple categorical and continuous variables, allowing incorporation of almost any kind of metadata).

-   Based on simulation studies, maintains the false discovery rate at or below the nominal rate, unlike some other packages.

-   Empirical Bayes smoothing of gene-wise standard deviations provides increased power.

### Basic Steps of Differential Gene Expression

1.  Read count data and annotation into R and preprocessing.

2.  Calculate normalization factors (sample-specific adjustments)

3.  Filter genes (uninteresting genes, e.g. unexpressed)

4.  Account for expression-dependent variability by transformation, weighting, or modeling

5.  Fitting a linear model

6.  Perform statistical comparisons of interest (using contrasts)

7.  Adjust for multiple testing, Benjamini-Hochberg (BH) or q-value

8.  Check results for confidence

9.  Attach annotation if available and write tables

```{r load_packages, echo=FALSE, warning=FALSE, message=FALSE}

library(edgeR)

library("RColorBrewer")

library("gplots")

```

## 1. Read in the counts table and create our DGEList

## Differential Gene Expression List Object (DGEList) object

A DGEList is an object in the package edgeR for storing count data, normalization factors, and other information

```{r dgelist}

d0 <- DGEList(counts.matrix)

```

**1a.** Read in Annotation

```{r read_annotation}

anno <- read.delim("ensembl_mm_106.tsv",as.is=T)

dim(anno)

head(anno)

tail(anno)

any(duplicated(anno$Gene.stable.ID))

```

**1b.** Derive experiment metadata from the sample names

Our experiment has two factors, genotype ("WT", "KOMIR150", or "KOTet3") and cell type ("C" or "NC").

The sample names are "mouse" followed by an animal identifier, followed by the genotype, followed by the cell type.

```{r metadata}

sample_names <- colnames(counts)

metadata <- as.data.frame(strsplit2(sample_names, c("_"))[,2:4], row.names = sample_names)

colnames(metadata) <- c("mouse", "genotype", "cell_type")

metadata

```

Create a new variable "group" that combines genotype and cell type.

```{r interaction}

metadata$group <- interaction(metadata$genotype, metadata$cell_type)

table(metadata$group)

table(metadata$mouse)

metadata

```

Note: you can also enter group information manually, or read it in from an external file. If you do this, it is $VERY, VERY, VERY$ important that you make sure the metadata is in the same order as the column names of the counts table.

## 2. Preprocessing and Normalization factors

In differential expression analysis, only sample-specific effects need to be normalized, we are NOT concerned with comparisons and quantification of absolute expression.

-   Sequence depth -- is a sample specific effect and needs to be adjusted for.

-   RNA composition - finding a set of scaling factors for the library sizes that minimize the log-fold changes between the samples for most genes (edgeR uses a trimmed mean of M-values between each pair of sample)

-   GC content -- is NOT sample-specific (except when it is)

-   Gene Length -- is NOT sample-specific (except when it is)

In edgeR/limma, you calculate normalization factors to scale the raw library sizes (number of reads) using the function calcNormFactors, which by default uses TMM (weighted trimmed means of M values to the reference). Assumes most genes are not DE.

Proposed by Robinson and Oshlack (2010).

```{r preprocess}

d0 <- calcNormFactors(d0)

d0$samples

```

**Note:** calcNormFactors doesn't *normalize* the data, it just calculates normalization factors for use downstream.

## 3. Filtering genes

We filter genes based on non-experimental factors to reduce the number of genes/tests being conducted and therefor do not have to be accounted for in our transformation or multiple testing correction. Commonly we try to remove genes that are either a) unexpressed, or b) unchanging (low-variability).

Common filters include:

1.  to remove genes with a max value (X) of less then Y.

2.  to remove genes that are less than X normalized read counts (cpm) across a certain number of samples. Ex: rowSums(cpms \<=1) \< 3 , require at least 1 cpm in at least 3 samples to keep.

3.  A less used filter is for genes with minimum variance across all samples, so if a gene isn't changing (constant expression) its inherently not interesting therefor no need to test.

We will use the built in function filterByExpr() to filter low-expressed genes. filterByExpr uses the experimental design to determine how many samples a gene needs to be expressed in to stay. Importantly, once this number of samples has been determined, the group information is not used in filtering.

Using filterByExpr requires specifying the model we will use to analysis our data.

-   The model you use will change for every experiment, and this step should be given the most time and attention.\*

We use a model that includes group and (in order to account for the paired design) mouse.

```{r model}

group <- metadata$group

mouse <- metadata$mouse

mm <- model.matrix(~0 + group + mouse)

head(mm)

```

```{r}

keep <- filterByExpr(d0, mm)

sum(keep) # number of genes retained

d <- d0[keep,]

```

"Low-expressed" depends on the dataset and can be subjective.

Visualizing your data with a Multidimensional scaling (MDS) plot.

```{r mds, fig.width=6}

plotMDS(d, col = as.numeric(metadata$group), cex=1)

```

The MDS plot tells you **A LOT** about what to expect from your experiment.

**3a.** Extracting "normalized" expression table

### RPKM vs. FPKM vs. CPM and Model Based

-   RPKM - Reads per kilobase per million mapped reads

-   FPKM - Fragments per kilobase per million mapped reads

-   logCPM -- log Counts per million [ good for producing MDS plots, estimate of normalized values in model based ]

-   Model based - original read counts are not themselves transformed, but rather correction factors are used in the DE model itself.

We use the `cpm` function with log=TRUE to obtain log-transformed normalized expression data. On the log scale, the data has less mean-dependent variability and is more suitable for plotting.

```{r cpm}

logcpm <- cpm(d, prior.count=2, log=TRUE)

write.table(logcpm,"rnaseq_workshop_normalized_counts.txt",sep="\t",quote=F)

```

## 4. Voom transformation and calculation of variance weights

**4a.** **Voom**

```{r voom, fig.width=6}

y <- voom(d, mm, plot = T)

```

What is voom doing?

1.  Counts are transformed to log2 counts per million reads (CPM), where "per million reads" is defined based on the normalization factors we calculated earlier.

2.  A linear model is fitted to the log2 CPM for each gene, and the residuals are calculated.

3.  A smoothed curve is fitted to the sqrt(residual standard deviation) by average expression.

(see red line in plot above)

4.  The smoothed curve is used to obtain weights for each gene and sample that are passed into limma along with the log2 CPMs.

More details at "[voom: precision weights unlock linear model analysis tools for RNA-seq read counts](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29)"

If your voom plot looks like the below (performed on the raw data), you might want to filter more:

```{r voom_bad, fig.width=6}

tmp <- voom(d0, mm, plot = T)

```

## 5. Fitting linear models in limma

lmFit fits a linear model using weighted least squares for each gene:

```{r lmfit}

fit <- lmFit(y, mm)

head(coef(fit))

```

Comparisons between groups (log fold-changes) are obtained as *contrasts* of these fitted linear models:

## 6. Specify which groups to compare using contrasts:

Comparison between cell types for genotype WT.

```{r contrast}

contr <- makeContrasts(groupWT.C - groupWT.NC, levels = colnames(coef(fit)))

contr

```

**6a.** Estimate contrast for each gene

```{r contrast.fit}

tmp <- contrasts.fit(fit, contr)

```

The variance characteristics of low expressed genes are different from high expressed genes, if treated the same, the effect is to over represent low expressed genes in the DE list. This is corrected for by the log transformation and voom. However, some genes will have increased or decreased variance that is not a result of low expression, but due to other random factors. We are going to run empirical Bayes to adjust the variance of these genes.

Empirical Bayes smoothing of standard errors (shifts standard errors that are much larger or smaller than those from other genes towards the average standard error) (see "[Linear Models and Empirical Bayes Methods for Assessing Differential Expression in Microarray Experiments](https://www.degruyter.com/doi/10.2202/1544-6115.1027)"

**6b.** Apply EBayes

```{r ebayes}

tmp <- eBayes(tmp)

```

## 7. Multiple Testing Adjustment

The TopTable. Adjust for multiple testing using method of Benjamini & Hochberg (BH), or its 'alias' fdr. "[Controlling the false discovery rate: a practical and powerful approach to multiple testing](http://www.jstor.org/stable/2346101).

here `n=Inf` says to produce the topTable for **all** genes.

```{r toptable}

top.table <- topTable(tmp, adjust.method = "BH", sort.by = "P", n = Inf)

```

### Multiple Testing Correction

Simply a must! Best choices are:

-   [FDR](http://www.jstor.org/stable/2346101) (false discovery rate), such as Benjamini-Hochberg (1995).

-   [Qvalue](https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/1467-9868.00346) - Storey (2002)

The FDR (or qvalue) is a statement about the list and is no longer about the gene (pvalue). So a FDR of 0.05, says you expect 5% false positives among the list of genes with an FDR of 0.05 or less.

The statement "Statistically significantly different" means FDR of 0.05 or less.

**7a.** How many DE genes are there (false discovery rate corrected)?

```{r count_de}

length(which(top.table$adj.P.Val < 0.05))

```

## 8. Check your results for confidence.

You've conducted an experiment, you've seen a phenotype. Now check which genes are most differentially expressed (show the top 50)? Look up these top genes, their description and ensure they relate to your experiment/phenotype.

```{r de_genes_top50}

head(top.table, 50)

```

Columns are

-   logFC: log2 fold change of WT.C/WT.NC

-   AveExpr: Average expression across all samples, in log2 CPM

-   t: logFC divided by its standard error

-   P.Value: Raw p-value (based on t) from test that logFC differs from 0

-   adj.P.Val: Benjamini-Hochberg false discovery rate adjusted p-value

-   B: log-odds that gene is DE (arguably less useful than the other columns)

ENSMUSG00000030203.18 has higher expression at WT NC than at WT C (logFC is negative). ENSMUSG00000026193.16 has higher expression at WT C than at WT NC (logFC is positive).

In the paper, the authors specify that NC cells were identified by low expression of Ly6C (which is now called Ly6c1 or ENSMUSG00000079018.11). Is this gene differentially expressed?

```{r}

top.table["ENSMUSG00000079018.11",]

d0$counts["ENSMUSG00000079018.11",]

```

Ly6c1 was removed from our data by the filtering step, because the maximum counts for the gene did not exceed 2.

## 9. Write top.table to a file, adding in cpms and annotation

```{r}

top.table$Gene <- rownames(top.table)

top.table <- top.table[,c("Gene", names(top.table)[1:6])]

top.table <- data.frame(top.table,anno[match(top.table$Gene,anno$Gene.stable.ID.version),],logcpm[match(top.table$Gene,rownames(logcpm)),])



head(top.table)

write.table(top.table, file = "WT.C_v_WT.NC.txt", row.names = F, sep = "\t", quote = F)

```

# Linear models and contrasts

Let's say we want to compare genotypes for cell type C. The only thing we have to change is the call to makeContrasts:

```{r}

contr <- makeContrasts(groupWT.C - groupKOMIR150.C, levels = colnames(coef(fit)))

tmp <- contrasts.fit(fit, contr)

tmp <- eBayes(tmp)

top.table <- topTable(tmp, sort.by = "P", n = Inf)

head(top.table, 20)

length(which(top.table$adj.P.Val < 0.05)) # number of DE genes

top.table$Gene <- rownames(top.table)

top.table <- top.table[,c("Gene", names(top.table)[1:6])]

top.table <- data.frame(top.table,anno[match(top.table$Gene,anno$Gene.stable.ID),],logcpm[match(top.table$Gene,rownames(logcpm)),])



write.table(top.table, file = "WT.C_v_KOMIR150.C.txt", row.names = F, sep = "\t", quote = F)

```

## What if we refit our model as a two-factor model (rather than using the group variable)?

Create new model matrix:

```{r two_factor}

genotype <- factor(metadata$genotype, levels = c("WT", "KOMIR150", "KOTet3"))

cell_type <- factor(metadata$cell_type, levels = c("C", "NC"))

mouse <- factor(metadata$mouse, levels = c("110", "148", "158", "183", "198", "206", "2670", "7530", "7531", "7532", "H510", "H514"))

mm <- model.matrix(~genotype*cell_type + mouse)

```

We are specifying that model includes effects for genotype, cell type, and the genotype-cell type interaction (which allows the differences between genotypes to differ across cell types).

```{r}

colnames(mm)

```

```{r}

y <- voom(d, mm, plot = F)

fit <- lmFit(y, mm)

head(coef(fit))

colnames(coef(fit))

```

-   The coefficient genotypeKOMIR150 represents the difference in mean expression between KOMIR150 and the reference genotype (WT), *for cell type C* (the reference level for cell type)

-   The coefficient cell_typeNC represents the difference in mean expression between cell type NC and cell type C, *for genotype WT*

-   The coefficient genotypeKOMIR150:cell_typeNC is the difference between cell types NC and C of the differences between genotypes KOMIR150 and WT (the interaction effect).

Let's estimate the difference between genotypes WT and KOMIR150 in cell type C.

```{r}

tmp <- contrasts.fit(fit, coef = 2) # Directly test second coefficient

tmp <- eBayes(tmp)

top.table <- topTable(tmp, sort.by = "P", n = Inf)

head(top.table, 20)

length(which(top.table$adj.P.Val < 0.05)) # number of DE genes

```

We get the same results as with the model where each coefficient corresponded to a group mean. In essence, these are the *same* model, so use whichever is most convenient for what you are estimating.

The interaction effects genotypeKOMIR150:cell_typeNC are easier to estimate and test in this setup.

```{r}

head(coef(fit))

colnames(coef(fit))

```

```{r}

tmp <- contrasts.fit(fit, coef = 16) # Test genotypeKOMIR150:cell_typeNC

tmp <- eBayes(tmp)

top.table <- topTable(tmp, sort.by = "P", n = Inf)

head(top.table, 20)

length(which(top.table$adj.P.Val < 0.05))

```

The log fold change here is the difference between genotypes KOMIR150 and WT in the log fold changes between cell types NC and C.

A gene for which this interaction effect is significant is one for which the effect of cell type differs between genotypes, and for which the effect of genotypes differs between cell types.

### More complicated models

Specifying a different model is simply a matter of changing the calls to model.matrix (and possibly to contrasts.fit).

What if we want to adjust for a continuous variable like some health score?

(We are making this data up here, but it would typically be a variable in your metadata.)

```{r}

# Generate example health data

set.seed(99)

HScore <- rnorm(n = 22, mean = 7.5, sd = 1)

HScore

```

Model adjusting for HScore score:

```{r}

mm <- model.matrix(~0 + group + mouse + HScore)

y <- voom(d, mm, plot = F)

fit <- lmFit(y, mm)

contr <- makeContrasts(groupKOMIR150.NC - groupWT.NC,

                       levels = colnames(coef(fit)))

tmp <- contrasts.fit(fit, contr)

tmp <- eBayes(tmp)

top.table <- topTable(tmp, sort.by = "P", n = Inf)

head(top.table, 20)

length(which(top.table$adj.P.Val < 0.05))

```

What if we want to look at the correlation of gene expression with a continuous variable like pH?

```{r}

# Generate example pH data

set.seed(99)

pH <- rnorm(n = 22, mean = 8, sd = 1.5)

pH

```

Specify model matrix:

```{r}

mm <- model.matrix(~pH)

head(mm)

```

```{r}

y <- voom(d, mm, plot = F)

fit <- lmFit(y, mm)

tmp <- contrasts.fit(fit, coef = 2) # test "pH" coefficient

tmp <- eBayes(tmp)

top.table <- topTable(tmp, sort.by = "P", n = Inf)

head(top.table, 20)

length(which(top.table$adj.P.Val < 0.05))

```

In this case, limma is fitting a linear regression model, which here is a straight line fit, with the slope and intercept defined by the model coefficients:

```{r}

ENSMUSG00000056054 <- y$E["ENSMUSG00000056054.10",]

plot(ENSMUSG00000056054 ~ pH, ylim = c(0, 3.5))

intercept <- coef(fit)["ENSMUSG00000056054.10", "(Intercept)"]

slope <- coef(fit)["ENSMUSG00000056054.10", "pH"]

abline(a = intercept, b = slope)

slope

```

In this example, the log fold change logFC is the slope of the line, or the change in gene expression (on the log2 CPM scale) for each unit increase in pH.

Here, a logFC of 0.20 means a 0.20 log2 CPM increase in gene expression for each unit increase in pH, or a 15% increase on the CPM scale (2\^0.20 = 1.15).

### A bit more on linear models

Limma fits a linear model to each gene.

Linear models include analysis of variance (ANOVA) models, linear regression, and any model of the form

Y = β<sub>0</sub> + β<sub>1</sub>X<sub>1</sub> + β<sub>2</sub>X<sub>2</sub> + ... + β<sub>p</sub>X<sub>p</sub> + ε

The covariates X can be:

-   a continuous variable (pH, HScore score, age, weight, temperature, etc.)

-   Dummy variables coding a categorical covariate (like cell type, genotype, and group)

The β's are unknown parameters to be estimated.

In limma, the β's are the log fold changes.

The error (residual) term ε is assumed to be normally distributed with a variance that is constant across the range of the data.

Normally distributed means the residuals come from a distribution that looks like this:

```{r, echo = F}

hist(rnorm(n = 10000), main = "Normally Distributed Data", breaks=50)

```

The log2 transformation that voom applies to the counts makes the data "normal enough", but doesn't completely stabilize the variance:

```{r}

mm <- model.matrix(~0 + group + mouse)

tmp <- voom(d, mm, plot = T)

```

The log2 counts per million are more variable at lower expression levels. The variance weights calculated by voom address this situation.

### Both edgeR and limma have VERY comprehensive user manuals

The limma users' guide has great details on model specification.

-   [Limma voom](https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf)

-   [edgeR](http://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf)

# Simple plotting

```{r}

mm <- model.matrix(~genotype*cell_type + mouse)

colnames(mm) <- make.names(colnames(mm))

y <- voom(d, mm, plot = F)

fit <- lmFit(y, mm)

contrast.matrix <- makeContrasts(genotypeKOMIR150, levels=colnames(coef(fit)))

fit2 <- contrasts.fit(fit, contrast.matrix)

fit2 <- eBayes(fit2)

top.table <- topTable(fit2, coef = 1, sort.by = "P", n = 40)

```

## Volcano plot

```{r fig.width=4, fig.height=4}

volcanoplot(fit2, coef=1, highlight=8, names=rownames(fit2), main="Genotype KOMIR150 vs. WT for cell type C", cex.main = 0.8)



head(anno[match(rownames(fit2), anno$Gene.stable.ID.version),

     c("Gene.stable.ID.version", "Gene.name") ])

identical(anno[match(rownames(fit2), anno$Gene.stable.ID.version),

     c("Gene.stable.ID.version")], rownames(fit2))



volcanoplot(fit2, coef=1, highlight=8, names=anno[match(rownames(fit2), anno$Gene.stable.ID.version), "Gene.name"], main="Genotype KOMIR150 vs. WT for cell type C", cex.main = 0.8)



```

## Heatmap

```{r fig.height=8}

#using a red and blue color scheme without traces and scaling each row

heatmap.2(logcpm[rownames(DEgene.trt.all),],col=brewer.pal(11,"RdBu"),scale="row", trace="none")



anno[match(rownames(DEgene.trt.all), anno$Gene.stable.ID.version),

     c("Gene.stable.ID.version", "Gene.name")]

identical(anno[match(rownames(top.table), anno$Gene.stable.ID.version), "Gene.stable.ID.version"], rownames(top.table))



heatmap.2(logcpm[rownames(DEgene.trt.all),],col=brewer.pal(11,"RdBu"),scale="row", trace="none", labRow = anno[match(rownames(top.table), anno$Gene.stable.ID.version), "Gene.name"])

```

## 2 factor venn diagram

```{r fig.width=6, fig.height=6}

mm <- model.matrix(~genotype*cell_type + mouse)

colnames(mm) <- make.names(colnames(mm))

y <- voom(d, mm, plot = F)

fit <- lmFit(y, mm)

contrast.matrix <- makeContrasts(genotypeKOMIR150, genotypeKOMIR150 + genotypeKOMIR150.cell_typeNC, levels=colnames(coef(fit)))

fit2 <- contrasts.fit(fit, contrast.matrix)

fit2 <- eBayes(fit2)

top.table <- topTable(fit2, coef = 1, sort.by = "P", n = 40)



results <- decideTests(fit2)

vennDiagram(results, names = c("C", "NC"), main = "DE Genes Between KOMIR150 and WT by Cell Type", cex.main = 0.8)

```

## Download the Enrichment Analysis R Markdown document

```{r download1, eval=FALSE}

download.file("https://raw.githubusercontent.com/ucdavis-bioinformatics-training/2020-mRNA_Seq_Workshop/master/data_analysis/enrichment_mm.Rmd", "enrichment_mm.Rmd")

```

```{r}

sessionInfo()

```
