---
title: "BradleyDEGs_revised"
author: "Britt"
date: "2024-10-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Assign data to variables:

```{r}
PM.counts <- read.table("Psme_RNASeq_July2024_Psme.1_0_counts.txt", header = TRUE) 
PM.counts

TC.counts <- read.table("Psme_RNASeq_July2024_Trucit1_1_counts.txt", header = TRUE) 
TC.counts
```
Filter counts by retaining only the reads that were over 10 count in at least three samples:

```{r}
TC.counts <- TC.counts[rowSums(TC.counts[,-1] > 10) >= 3,]
TC.counts

PM.counts <- PM.counts[rowSums(PM.counts[,-1] > 10) >= 3,]
PM.counts
```

Remove Transcript ID column (convert dataframe to tibble):

```{r}
library(tibble)

TC.counts.noid <- as_tibble(TC.counts)
TC_counts_noid

PM.counts.noid <- as_tibble(PM.counts)
PM.counts.noid
```

Log transform the counts data:

```{r}
transformed.TC.counts <- log2(TC.counts.noid + 1) 
transformed.PM.counts <- log2(PM.counts.noid + 1)
```

Make tibble with descriptions from counts data table: (this time treating TC or ST as both just "M" for simpler statistical analysis)

```{r}
sample.description <- tibble(sample=colnames(TC.counts.noid))
sample.description <- sample.description %>%
  mutate(
    sp = c("NM", "NM", "NM", "NM", "M", "NM", "M", "M", "M", "NM", "M", "NM", "NM", "M", "NM", "M", "NM", "NM", "NM", "M", "NA"),  
    trt = c("C", "C", "T", "C", "C", "T", "T", "T", "C", "C", "C", "C", "T", "T", "C", "C", "C", "T", "T", "T", "NA"),  
    group = paste(sp, trt, sep = "_")
  )
sample.description <- sample.description %>%
  slice(1:(n() - 1))
sample.description <- sample.description %>%
  mutate(sp=factor(sp, levels = c("NM","M")), 
         trt=factor(trt,levels = c("C","T"))) # setting the levels in this way makes "C and NM" the reference
sample.description
```








