---
title: "Bradley DEGs"
author: "Britt"
date: "2024-09-03"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Download counts tables from Bogar server:

```{bash, eval = FALSE}
#in personal computer terminal:

scp -P 22345 bklong@bogarprecision.plb.ucdavis.edu:~/DemRNAanalysis/BioShare/04-Counts/Psme_RNASeq_July2024_Psme.1_0_counts.txt /c/Users/britt/Downloads

scp -P 22345 bklong@bogarprecision.plb.ucdavis.edu:~/DemRNAanalysis/BioShare/04-Counts/Psme_RNASeq_July2024_Trucit1_1_counts.txt /c/Users/britt/Downloads
```

The counts tables are now on my personal computer where R can access them.

**Basic steps for DGE Analysis:**

1.  Load the RNAseq counts.
2.  Normalize the counts.
3.  QC the counts.
4.  Create a data frame that describes the experiment.
5.  Determine the dispersion parameter (how over-dispersed is the data?)
6.  Fit a statistical model for gene expression as a function of experimental parameters.
7.  Test the significance of experimental parameters for explaining gene expression.
8.  Examine result.

Let's take a look at the data:

```{r}
PM_counts <- read.table("Psme_RNASeq_July2024_Psme.1_0_counts.txt", header = TRUE) 

PM_counts
```

```{r}
TC_counts <- read.table("Psme_RNASeq_July2024_Trucit1_1_counts.txt", header = TRUE) 
TC_counts
```

Filter the counts data so that you only retain genes that were expressed:

```{r}
#retain only the reads that were over 10 count in at least three samples

TC_counts <- TC_counts[rowSums(TC_counts[,-1] > 10) >= 3,]

TC_counts

PM_counts <- PM_counts[rowSums(PM_counts[,-1] > 10) >= 3,]

PM_counts
```

Remove Transcript ID column (convert dataframe to tibble).

```{r}
library(tibble)

TC_counts_noid <- as_tibble(TC_counts)

TC_counts_noid

PM_counts_noid <- as_tibble(PM_counts)

PM_counts_noid
```

Log transform the counts data.

```{r}
transformed.TC.counts <- log2(TC_counts_noid + 1) 
transformed.PM.counts <- log2(PM_counts_noid + 1)
```

Correlation table and pairs:

```{r}
cor.table.TC <- cor(TC_counts_noid)
cor.table.PM <- cor(PM_counts_noid)
```

```{r}
pairs(transformed.PM.counts)
pairs(transformed.TC.counts)
```

```{r}
library(tidyverse)

cor.table.PM %>% gplots::heatmap.2(dendrogram="row", trace = "none", col=viridis::viridis(25, begin=.25), margins=c(4,4))
```

```{r}
cor.table.TC %>% gplots::heatmap.2(dendrogram="row", trace = "none", col=viridis::viridis(25, begin=.25), margins=c(4,4))
```

```{r}
#make tibble with descriptions from counts data table
sample.description <- tibble(sample=colnames(TC_counts_noid))

sample.description
```

```{r}
sample.description <- sample.description %>%
  mutate(
    sp = c("NM", "NM", "NM", "NM", "ST", "NM", "TC", "ST", "ST", "NM", "TC", "NM", "NM", "ST", "NM", "ST", "NM", "NM", "NM", "ST", "NA"),  
    trt = c("C", "C", "T", "C", "C", "T", "T", "T", "C", "C", "C", "C", "T", "T", "C", "C", "C", "T", "T", "T", "NA"),  
    group = paste(sp, trt, sep = "_")
  )

sample.description
```

```{r}
sample.description <- sample.description %>%
  slice(1:(n() - 1))
sample.description
```
