---
title: "BradleyDEGs_revised"
author: "Britt"
date: "2024-10-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Assign data to variables:

```{r}
PM.counts <- read.table("Psme_RNASeq_July2024_Psme.1_0_counts.txt", header = TRUE) 
PM.counts

TC.counts <- read.table("Psme_RNASeq_July2024_Trucit1_1_counts.txt", header = TRUE) 
TC.counts
```
Filter counts by retaining only the reads that were over 10 count in at least three samples:

```{r}
TC.counts <- TC.counts[rowSums(TC.counts[,-1] > 10) >= 3,]
TC.counts

PM.counts <- PM.counts[rowSums(PM.counts[,-1] > 10) >= 3,]
PM.counts
```

Remove Transcript ID column (convert dataframe to tibble):

```{r}
library(tibble)

TC.counts.noid <- as_tibble(TC.counts)
TC_counts_noid

PM.counts.noid <- as_tibble(PM.counts)
PM.counts.noid
```

Log transform the counts data:

```{r}
transformed.TC.counts <- log2(TC.counts.noid + 1) 
transformed.PM.counts <- log2(PM.counts.noid + 1)
```

Make tibble with descriptions from counts data table: (this time treating TC or ST as both just "M" for simpler statistical analysis)

```{r}
sample.description <- tibble(sample=colnames(TC.counts.noid))
sample.description <- sample.description %>%
  mutate(
    sp = c("NM", "NM", "NM", "NM", "M", "NM", "M", "M", "M", "NM", "M", "NM", "NM", "M", "NM", "M", "NM", "NM", "NM", "M", "NA"),  
    trt = c("C", "C", "T", "C", "C", "T", "T", "T", "C", "C", "C", "C", "T", "T", "C", "C", "C", "T", "T", "T", "NA"),  
    group = paste(sp, trt, sep = "_")
  )
sample.description <- sample.description %>%
  slice(1:(n() - 1))
sample.description <- sample.description %>%
  mutate(sp=factor(sp, levels = c("NM","M")), 
         trt=factor(trt,levels = c("C","T"))) # setting the levels in this way makes "C and NM" the reference
sample.description
```

DGE Analysis:

```{r}
library(edgeR)

counts.matrix.TC <- TC.counts %>% as.matrix()
counts.matrix.PM <- PM.counts %>% as.matrix()

#remove unassigned row
counts.matrix.TC <- counts.matrix.TC[, -21]
counts.matrix.PM <- counts.matrix.PM[, -21]

counts.matrix.TC
counts.matrix.PM
```

```{r}
dge.data.TC <- DGEList(counts=counts.matrix.TC,
                       group=sample.description$group)
dim(dge.data.TC)
dge.data.TC <- calcNormFactors(dge.data.TC, method = "TMM")
dge.data.TC$samples #look at the normalization factors

dge.data.PM <- DGEList(counts=counts.matrix.PM, 
                    group=sample.description$group)
dim(dge.data.PM) 
dge.data.PM <- calcNormFactors(dge.data.PM, method = "TMM")
dge.data.PM$samples # look at the normalization factors
```

MDS plot:

```{r}
mdsvals <- plotMDS(dge.data.TC, plot = FALSE) # get the MDS values for plotting

tibble(x=mdsvals$x, y=mdsvals$y, sample=rownames(dge.data.TC$samples)) %>%
  inner_join(sample.description) %>%
  ggplot(aes(x=x, y=y, color=trt, shape=sp)) +
  geom_point(size=3) 
```
```{r}
#sample.description is the original descriptive dataset (describes our samples but doesn't have counts data)

design.TC <- model.matrix(~sp+trt,data=sample.description)
rownames(design.TC) <- sample.description$sample
design.TC

design.PM <- model.matrix(~sp+trt,data = sample.description)
rownames(design.PM) <- sample.description$sample
design.PM
```
```{r}
#First the overall dispersion
dge.data.TC <- estimateGLMCommonDisp(dge.data.TC, design.TC, verbose = TRUE)
dge.data.PM <- estimateGLMCommonDisp(dge.data.PM,design.PM,verbose = TRUE)

#Then a trended dispersion based on count level
dge.data.TC <- estimateGLMCommonDisp(dge.data.TC,design.TC)
dge.data.PM <- estimateGLMTrendedDisp(dge.data.PM,design.PM)

#And lastly we calculate the gene-wise dispersion, using the prior estimates to "squeeze" the dispersion towards the common dispersion.
dge.data.TC <- estimateGLMTagwiseDisp(dge.data.TC,design.TC)
dge.data.PM <- estimateGLMTagwiseDisp(dge.data.PM,design.PM)
```
```{r}
#this is the full model with all coefficients
#dge.data is the normalized count data
fit.TC <- glmFit(dge.data.TC, design.TC)
fit.PM <- glmFit(dge.data.PM, design.PM)

#this model looks at genes differentially expressed between two treatments
trt.lrt.TC <- glmLRT(fit.TC, coef = "trtT")
trt.lrt.PM <- glmLRT(fit.PM,coef = "trtT")
sp.lrt.PM <- glmLRT(fit.PM, coef = "spM")
```

top 10 DEGs:

```{r}
top10.trt.TC <- topTags(trt.lrt.TC)
top10.trt.PM <- topTags(trt.lrt.PM)
top10.sp.PM <- topTags(sp.lrt.PM) 

top10.trt.TC
top10.trt.PM
top10.sp.PM
```






