---
title: "Bradley DEGs"
author: "Britt"
date: "2024-09-03"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Download counts tables from Bogar server:

```{bash, eval = FALSE}
#in personal computer terminal:

scp -P 22345 bklong@bogarprecision.plb.ucdavis.edu:~/DemRNAanalysis/BioShare/04-Counts/Psme_RNASeq_July2024_Psme.1_0_counts.txt /c/Users/britt/Downloads

scp -P 22345 bklong@bogarprecision.plb.ucdavis.edu:~/DemRNAanalysis/BioShare/04-Counts/Psme_RNASeq_July2024_Trucit1_1_counts.txt /c/Users/britt/Downloads
```

The counts tables are now on my personal computer where R can access them.

**Basic steps for DGE Analysis:**

1.  Load the RNAseq counts.
2.  Normalize the counts.
3.  QC the counts.
4.  Create a data frame that describes the experiment.
5.  Determine the dispersion parameter (how over-dispersed is the data?)
6.  Fit a statistical model for gene expression as a function of experimental parameters.
7.  Test the significance of experimental parameters for explaining gene expression.
8.  Examine result.

Let's take a look at the data:

```{r}
PM_counts <- read.table("Psme_RNASeq_July2024_Psme.1_0_counts.txt", header = TRUE) 

PM_counts
```

```{r}
TC_counts <- read.table("Psme_RNASeq_July2024_Trucit1_1_counts.txt", header = TRUE) 
TC_counts
```

Filter the counts data so that you only retain genes that were expressed:

```{r}
#retain only the reads that were over 10 count in at least three samples

TC_counts <- TC_counts[rowSums(TC_counts[,-1] > 10) >= 3,]

TC_counts

PM_counts <- PM_counts[rowSums(PM_counts[,-1] > 10) >= 3,]

PM_counts
```
