---
title: "BradleyDEGs_revised"
author: "Britt"
date: "2024-10-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Assign data to variables:

```{r}
PM.counts <- read.table("Psme_RNASeq_July2024_Psme.1_0_counts.txt", header = TRUE) 
PM.counts

TC.counts <- read.table("Psme_RNASeq_July2024_Trucit1_1_counts.txt", header = TRUE) 
TC.counts
```
Filter counts by retaining only the reads that were over 10 count in at least three samples:

```{r}
TC.counts <- TC.counts[rowSums(TC.counts[,-1] > 10) >= 3,]
TC.counts

PM.counts <- PM.counts[rowSums(PM.counts[,-1] > 10) >= 3,]
PM.counts
```

Remove Transcript ID column (convert dataframe to tibble):

```{r}
library(tibble)

TC.counts.noid <- as_tibble(TC.counts)
TC.counts.noid

PM.counts.noid <- as_tibble(PM.counts)
PM.counts.noid
```

Log transform the counts data: (I am not sure where this is used or if it's needed.)

```{r}
transformed.TC.counts <- log2(TC.counts.noid + 1) 
transformed.PM.counts <- log2(PM.counts.noid + 1)
```

Make tibble with descriptions from counts data table: (this time treating TC or ST as both just "M" for simpler statistical analysis)

```{r}
library(tidyverse)
sample.description <- tibble(sample=colnames(TC.counts.noid))
sample.description <- sample.description %>%
  mutate(
    sp = c("NM", "NM", "NM", "NM", "M", "NM", "M", "M", "M", "NM", "M", "NM", "NM", "M", "NM", "M", "NM", "NM", "NM", "M", "NA"),  
    trt = c("C", "C", "T", "C", "C", "T", "T", "T", "C", "C", "C", "C", "T", "T", "C", "C", "C", "T", "T", "T", "NA"),  
    group = paste(sp, trt, sep = "_")
  )
sample.description <- sample.description %>%
  slice(1:(n() - 1))
sample.description <- sample.description %>%
  mutate(sp=factor(sp, levels = c("NM","M")), 
         trt=factor(trt,levels = c("C","T"))) # setting the levels in this way makes "C and NM" the reference
sample.description
```

DGE Analysis:

```{r}
library(edgeR)

counts.matrix.TC <- TC.counts %>% as.matrix()
counts.matrix.PM <- PM.counts %>% as.matrix()

#remove unassigned row
counts.matrix.TC <- counts.matrix.TC[, -21]
counts.matrix.PM <- counts.matrix.PM[, -21]

counts.matrix.TC
counts.matrix.PM
```

```{r}
dge.data.TC <- DGEList(counts=counts.matrix.TC,
                       group=sample.description$group)
dim(dge.data.TC)
dge.data.TC <- calcNormFactors(dge.data.TC, method = "TMM")
dge.data.TC$samples #look at the normalization factors

dge.data.PM <- DGEList(counts=counts.matrix.PM, 
                    group=sample.description$group)
dim(dge.data.PM) 
dge.data.PM <- calcNormFactors(dge.data.PM, method = "TMM")
dge.data.PM$samples # look at the normalization factors
```

MDS plot:

```{r}
mdsvals <- plotMDS(dge.data.TC, plot = FALSE) # get the MDS values for plotting

tibble(x=mdsvals$x, y=mdsvals$y, sample=rownames(dge.data.TC$samples)) %>%
  inner_join(sample.description) %>%
  ggplot(aes(x=x, y=y, color=trt, shape=sp)) +
  geom_point(size=3) 
```
```{r}
mdsvalsPM <- plotMDS(dge.data.PM.interaction, plot = FALSE) # get the MDS values for plotting

tibble(x=mdsvalsPM$x, y=mdsvalsPM$y, sample=rownames(dge.data.PM.interaction$samples)) %>%
  inner_join(sample.description) %>%
  ggplot(aes(x=x, y=y, color=trt, shape=sp)) +
  geom_point(size=3) 
```


```{r}
mdsvals <- plotMDS(dge.data.PM, plot = FALSE) #using this one for paper

var1 <- round(mdsvals$var.explained[1] * 100, 1)
var2 <- round(mdsvals$var.explained[2] * 100, 1)

tibble(
  x = mdsvals$x,
  y = mdsvals$y,
  sample = rownames(dge.data.PM$samples)
) %>%
  inner_join(sample.description, by = "sample") %>%
  ggplot(aes(x = x, y = y, color = trt, shape = sp)) +
  geom_point(size = 3) +
  labs(
    x = paste0("Dimension 1 (", var1, "%)"),
    y = paste0("Dimension 2 (", var2, "%)"),
    color = "Drought",
    shape = "Mycorrhization"
  ) +
  theme_classic()
```



```{r}
#sample.description is the original descriptive dataset (describes our samples but doesn't have counts data)

design.TC <- model.matrix(~sp+trt,data=sample.description)
rownames(design.TC) <- sample.description$sample
design.TC

design.PM <- model.matrix(~sp+trt,data = sample.description)
rownames(design.PM) <- sample.description$sample
design.PM

design.PM.interaction <- model.matrix(~sp*trt,data = sample.description)
rownames(design.PM.interaction) <- sample.description$sample
design.PM.interaction
```
```{r}
#First the overall dispersion
dge.data.TC <- estimateGLMCommonDisp(dge.data.TC, design.TC, verbose = TRUE)
dge.data.PM <- estimateGLMCommonDisp(dge.data.PM,design.PM,verbose = TRUE)
dge.data.PM.interaction <- estimateGLMCommonDisp(dge.data.PM,design.PM.interaction,verbose = TRUE)

#Then a trended dispersion based on count level
dge.data.TC <- estimateGLMCommonDisp(dge.data.TC,design.TC)
dge.data.PM <- estimateGLMTrendedDisp(dge.data.PM,design.PM)
dge.data.PM.interaction <- estimateGLMTrendedDisp(dge.data.PM.interaction,design.PM.interaction)

#And lastly we calculate the gene-wise dispersion, using the prior estimates to "squeeze" the dispersion towards the common dispersion.
dge.data.TC <- estimateGLMTagwiseDisp(dge.data.TC,design.TC)
dge.data.PM <- estimateGLMTagwiseDisp(dge.data.PM,design.PM)
dge.data.PM.interaction <- estimateGLMTagwiseDisp(dge.data.PM.interaction,design.PM.interaction)
```
```{r}
#this is the full model with all coefficients
#dge.data is the normalized count data
fit.TC <- glmFit(dge.data.TC, design.TC)
fit.PM <- glmFit(dge.data.PM, design.PM)
fit.interaction <- glmFit(dge.data.PM.interaction, design.PM.interaction)

#this model looks at genes differentially expressed between two treatments
trt.lrt.TC <- glmLRT(fit.TC, coef = "trtT")
trt.lrt.PM <- glmLRT(fit.PM,coef = "trtT")
sp.lrt.PM <- glmLRT(fit.PM, coef = "spM")
interaction.lrt.PM <- glmLRT(fit.interaction,coef = "spM:trtT")
```

top 10 DEGs:

```{r}
top10.trt.TC <- topTags(trt.lrt.TC)
top10.trt.PM <- topTags(trt.lrt.PM)
top10.sp.PM <- topTags(sp.lrt.PM) 
top10.interaction.PM <- topTags(interaction.lrt.PM)

top10.trt.TC
top10.trt.PM
top10.sp.PM
top10.interaction.PM
```
summary:

```{r}
summary(decideTestsDGE(trt.lrt.TC,p.value=0.01)) #This uses the FDR.  0.05 would be OK also.
summary(decideTestsDGE(trt.lrt.PM,p.value=0.01)) 
summary(decideTestsDGE(sp.lrt.PM,p.value=0.01))
summary(decideTestsDGE(interaction.lrt.PM,p.value=0.05)) 
```
To summarize above analysis: No significant DEGs found between treatments for the fungal counts (at both 0.01 and 0.05 FDR). Significant DEGs were found (FDR 0.01) between treatments and species for tree counts. The interaction analysis shows no significant DEGs at FDR 0.01 but that increases to 29 significant DEGs at FDR 0.05. These 29 interaction genes are of high interest to our study. The next step is to try and look at the function of these, as well as create visualizations of the data.

Let's make a heat map...

Get normalized and transformed counts:
```{r}
PM.counts.normal.log <- cpm(dge.data.PM,log = TRUE)
TC.counts.normal.log <- cpm(dge.data.TC,log = TRUE)
```

Make basic heatmap:
```{r}
library(pheatmap)
pheatmap(PM.counts.normal.log, scale = "row")
pheatmap(TC.counts.normal.log, scale = "row")
```

make new dataframe to make heatmap that has treatments included in sample labels:

```{r}
PM.counts.normal.log.named <- PM.counts.normal.log
colnames(PM.counts.normal.log.named) <- c("NM_C_001", "NM_C_005", "NM_T_009", "NM_C_016", "M_C_019", "NM_T_023", "M_T_026", "M_T_027", "M_C_030", "NM_C_043", "M_C_045", "NM_C_049", "NM_T_050", "M_T_052", "NM_C_057", "M_C_062", "NM_C_080", "NM_T_082", "NM_T_085", "M_T_087")

PM.counts.normal.log.named
```
```{r}
TC.counts.normal.log.named <- TC.counts.normal.log[, c("DFD_019", "DFD_026", "DFD_027", "DFD_030", "DFD_045", "DFD_052", "DFD_062", "DFD_087")]

colnames(TC.counts.normal.log.named) <- c("C_019", "T_026", "T_027", "C_030", "C_045", "T_052","C_062", "T_087")

TC.counts.normal.log.named 
```
```{r}
pheatmap(PM.counts.normal.log.named, scale = "row")
pheatmap(TC.counts.normal.log.named, scale = "row")
```
subset top genes for heatmap:

```{r}
top50.PM.trt <- rownames(topTags(trt.lrt.PM, n = 50)$table)
top50.PM.trt.named <- PM.counts.normal.log.named[top50.PM.trt, ]

pheatmap(top50.PM.trt.named, scale = "row")
```
```{r}
top50.PM.sp <- rownames(topTags(sp.lrt.PM, n = 50)$table)
top50.PM.sp.named <- PM.counts.normal.log.named[top50.PM.sp, ]

pheatmap(top50.PM.sp.named, scale = "row")
```
```{r}
top15.PM.sp <- rownames(topTags(sp.lrt.PM, n = 15)$table)
top15.PM.sp.named <- PM.counts.normal.log.named[top15.PM.sp, ]

pheatmap(top15.PM.sp.named, scale = "row")
```


```{r}
top50.TC.trt <- rownames(topTags(trt.lrt.TC, n = 50)$table)
top50.TC.trt.named <- TC.counts.normal.log.named[top50.TC.trt, ]

pheatmap(top50.TC.trt.named, scale = "row")
```

```{r}
top50.PM.interaction <- rownames(topTags(interaction.lrt.PM, n = 50)$table)
top50.PM.interaction.named <- PM.counts.normal.log.named[top50.PM.interaction, ]

pheatmap(top50.PM.interaction.named, scale = "row")
```
```{r}
top10.PM.interaction <- rownames(topTags(interaction.lrt.PM, n = 10)$table)
top10.PM.interaction.named <- PM.counts.normal.log.named[top10.PM.interaction, ]

pheatmap(top10.PM.interaction.named, scale = "row")
```
```{r}
top29.PM.interaction <- rownames(topTags(interaction.lrt.PM, n = 29)$table)
top29.PM.interaction.named <- PM.counts.normal.log.named[top29.PM.interaction, ]

pheatmap(top29.PM.interaction.named, scale = "row")
```

let's try making a volcano plot:

```{r}
# Typically contains columns: logFC, PValue, and FDR
library(ggplot2)
library(dplyr)

# Extract the DEGs from edgeR results properly
deg_results <- topTags(interaction.lrt.PM, n = Inf)$table %>%
  as.data.frame() %>%  # Ensure it's a proper data frame
  mutate(
    neg_log10_pval = -log10(PValue),  # Convert p-value to -log10 scale
    significance = case_when(
      FDR < 0.05 & abs(logFC) > 0.5 ~ "Significant",  # Adjust thresholds as needed
      TRUE ~ "Not Significant"
    )
  )

# Volcano plot
ggplot(deg_results, aes(x = logFC, y = neg_log10_pval, color = significance)) +
  geom_point(alpha = 0.6) +
  scale_color_manual(values = c("Significant" = "red", "Not Significant" = "gray")) +
  theme_minimal() +
  labs(
    x = "Log2 Fold Change",
    y = "-Log10 P-value",
    title = "Volcano Plot of Differential Gene Expression"
  ) +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +  # FDR threshold
  geom_vline(xintercept = c(-0.5, 0.5), linetype = "dashed")  # Fold-change cutoff


```
let's adjust our heatmap to label significant genes:

```{r}
deg_results <- topTags(interaction.lrt.PM, n = Inf)$table %>%
  as.data.frame() %>%
  rownames_to_column(var = "Gene") %>%  # Convert rownames to a column
  mutate(
    neg_log10_pval = -log10(FDR),  
    significance = case_when(
      FDR < 0.05 & abs(logFC) > 0.5 ~ "Significant (FDR < 0.05)",
      TRUE ~ "Not Significant"
    ),
    gene_label = ifelse(significance == "Significant (FDR < 0.05)", Gene, NA)  # Use new Gene column
  )

library(ggrepel)


# Volcano plot
ggplot(deg_results, aes(x = logFC, y = neg_log10_pval, color = significance)) +
  geom_point(alpha = 0.6) +
  geom_text_repel(
    data = deg_results %>% filter(Gene == "PSME_21469"),  
    aes(label = "Cytochrome P450"), 
    max.overlaps = 15,
    size = 3
  ) +
  scale_color_manual(values = c("Significant (FDR < 0.05)" = "red", "Not Significant" = "gray")) +
  theme_minimal() +
  labs(
    x = "Log2 Fold Change",
    y = "-Log10 FDR",
  ) +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +  # FDR threshold
  geom_vline(xintercept = c(-0.5, 0.5), linetype = "dashed")  # Fold-change cutoff

```
```{r}
# Volcano plot
ggplot(deg_results, aes(x = logFC, y = neg_log10_pval, color = significance)) +
  geom_point(alpha = 0.6) +
  scale_color_manual(values = c("Significant (FDR < 0.05)" = "red", "Not Significant" = "gray")) +
  theme_minimal() +
  labs(
    x = "Log2 Fold Change",
    y = "-Log10 FDR"
  ) +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") +  # FDR threshold
  geom_vline(xintercept = c(-0.5, 0.5), linetype = "dashed")    # Fold-change cutoff


```


let's subset some genes to start looking at their function:

```{r}
#Extract genes with a FDR < 0.05
PM.interaction.DEGs <- topTags(interaction.lrt.PM,n = Inf,p.value = 0.05)$table

#save to a file
write.csv(PM.interaction.DEGs,"PM.interaction.DEGs.csv")
```

I will now make a new doc to use this dataframe for gene function stuff

----------------------------------------------------------------------

4/21/25

Let's try to filter PSME DEGs so that I can compare the 29 sig DEGs in the interaction treatment to the T/C and M/NM treatments alone. 

```{r}
print(PM.interaction.DEGs)
```

let's move rownames to a column (make a column of gene names)
```{r}
library(tibble)

PM.interaction.DEGs.col <- rownames_to_column(PM.interaction.DEGs, var = "Gene")
PM.interaction.DEGs.col
```

now let's extract our 29 genes of interest from this df:
```{r}
genes_of_interest <- PM.interaction.DEGs.col$Gene  
genes_of_interest
```

perfect, before we can start filtering for these genes in the other two treatments, we need dfs that contain ALL gene expression for those two treatments:
```{r}
all.genes.PM.trt <- topTags(trt.lrt.PM, n = Inf)$table
all.genes.PM.sp <- topTags(sp.lrt.PM, n = Inf)$table

all.genes.PM.trt.col <- rownames_to_column(all.genes.PM.trt, var = "Gene")
all.genes.PM.trt.col
all.genes.PM.sp.col <- rownames_to_column(all.genes.PM.sp, var = "Gene")
all.genes.PM.sp.col
```


let's try filtering for these genes only in the other two treatment scenarios:
```{r}
filtered.genes.PM.trt <- all.genes.PM.trt.col %>%
  filter(Gene %in% genes_of_interest)
filtered.genes.PM.sp <- all.genes.PM.sp.col %>%
  filter(Gene %in% genes_of_interest)

filtered.genes.PM.sp
filtered.genes.PM.trt
```

compare side-by-side:
```{r}

PM.interaction.DEGs.comparison <- PM.interaction.DEGs.col %>%
  inner_join(filtered.genes.PM.trt, by = "Gene", suffix = c("_interaction", "_trt")) %>%
  inner_join(filtered.genes.PM.sp, by = "Gene")

PM.interaction.DEGs.comparison
```

let's retain only the LogFC and FDR columns for ease of comparison:
```{r}
comparison <- PM.interaction.DEGs.comparison %>%
  select(Gene, starts_with("logFC"), starts_with("FDR"))
comparison
```

let's make a heatmap to compare expression across treatments!
```{r}
logfc_matrix <- comparison %>%
  column_to_rownames("Gene") %>%
  select(starts_with("logFC")) %>%
  as.matrix()

# Rename treatment columns to more intuitive labels
colnames(logfc_matrix) <- c("Interaction", "Drought", "Mycorrhization")

```

heatmap:
```{r}
library(pheatmap)
library(viridis)

pheatmap(logfc_matrix,
         cluster_rows = TRUE,
         cluster_cols = TRUE,
         scale = "none",  # or "row" for z-scores per gene
         color = colorRampPalette(c("blue", "white", "orange"))(100),
         main = "PSME Gene Expression: Log Fold Change Across Treatments",
         fontsize_row = 8,  # Adjust font size for row labels
         angle_col = 0)    # Rotate column labels to 0° (horizontal)

```

let's try naming the genes on the y axis

```{r}
rownames(logfc_matrix)

new_gene_names <- c("21469: Cytochrome P450", "27306: dormancy/auxin-related", "31667: GMC oxidoreductase", "38503: CRK", "30375: R13L4/SHOC-2-like", "29048: DUF829", "51370: GMC oxidoreductase", "27294: FAD/NAD(P)-binding monooxygenase", "23020: unknown", "37466: Myc-like bHLH", "13638: DCN1-like", "42174: siRNA processing", "19388: UDP-glycosyltransferase", "51302: UspA", "06003: MKZ1-like", "12596: myo-insitol-1-phosphate synthase", "16206: Fe(2+)/2OG dioxygenase", "47136: XTH", "12910: pectinesterase inhibitor", "04115: CRISP", "04789: sulfotransferase", "31428: TauD/TfdA-like", "11752: MFS transporter", "08506: JINGUBANG-like", "39647: CRK", "48863: MFS transporter", "40178: BURP domain", "38415: UspA", "40947: GH16")

rownames(logfc_matrix) <- new_gene_names

```

```{r}
pheatmap(logfc_matrix,
         cluster_rows = TRUE,
         cluster_cols = TRUE,
         scale = "none",  # or "row" for z-scores per gene
         color = colorRampPalette(c("blue", "white", "orange"))(100),
         fontsize_row = 8,  # Adjust font size for row labels
         angle_col = 0)    # Rotate column labels to 0° (horizontal)
```

We're going to attempt an upset plot:

```{r}
library(UpSetR)
library(ComplexUpset)
```

filter DEGs for FDR value:
```{r}
interaction.FDR.0.05 <- topTags(interaction.lrt.PM, n = Inf, p.value = 0.05)$table
sp.FDR.0.05 <- topTags(sp.lrt.PM, n = Inf, p.value = 0.05)$table
trt.FDR.0.05 <- topTags(trt.lrt.PM, n = Inf, p.value = 0.05)$table
```


make vectors of genes in different treatment groupings:
```{r}
sp.degs <- rownames(sp.FDR.0.05) 
trt.degs <- rownames(trt.FDR.0.05)
interaction.degs <- rownames(interaction.FDR.0.05)
```

make list of gene sets:
```{r}
gene.sets <- list(
  Mycorrhization = sp.degs,
  Drought = trt.degs,
  Interaction = interaction.degs
)
```


make plot:
```{r}
library(ComplexUpset)
```

```{r}
gene.matrix <- fromList(gene.sets)
```


```{r}
upset(
  gene.matrix,
  intersect = c("Mycorrhization", "Drought", "Interaction")
) +
labs(x = "Treatment Combination")
```
let's make a table of data comparing LogFC and FDR in 3 treatment groups...
```{r}
library(knitr)
```

```{r}
kable(comparison)


```


```{r}
write.csv((comparison), "comparison.csv", row.names = FALSE)
```

let's try to analyze interpro outputs:

```{r}
# Load necessary library
library(dplyr)
library(readr)

# Read InterPro file
interpro <- read_tsv("DFD/InterPro/PSME_04789_interpro.tsv", 
                     col_names = FALSE, comment = "#")

# Assign standard column names
colnames(interpro) <- c(
  "Protein_Accession", "Sequence_MD5", "Sequence_Length", "Analysis", 
  "Signature_Accession", "Signature_Description", "Start", "End", 
  "Score", "Status", "Date", "InterPro_Accession", 
  "InterPro_Description", "GO_Terms", "Pathways"
)

# Filter to rows with GO terms and InterPro descriptions
summary_df <- interpro %>%
  filter(GO_Terms != "-", InterPro_Description != "-") %>%
  group_by(Protein_Accession) %>%
  summarise(
    InterPro_Accession = paste(unique(InterPro_Accession), collapse = "; "),
    InterPro_Description = paste(unique(InterPro_Description), collapse = "; "),
    GO_Terms = paste(unique(GO_Terms), collapse = "; ")
  )

summary_df

```


```{r}
# Write to CSV
write_csv(summary_df, "interpro_GO_summary.csv")
```


```{r}
files <- list.files("DFD/InterPro", pattern = "_interpro.tsv$", full.names = TRUE)
summary_list <- list()

for (f in files) {
  interpro <- read_tsv(f, col_names = FALSE, comment = "#")
  colnames(interpro) <- c(
    "Protein_Accession", "Sequence_MD5", "Sequence_Length", "Analysis", 
    "Signature_Accession", "Signature_Description", "Start", "End", 
    "Score", "Status", "Date", "InterPro_Accession", 
    "InterPro_Description", "GO_Terms", "Pathways"
  )
  
  # Summarize here inside the loop, replacing "-" with NA or keeping all proteins
  summary_df <- interpro %>%
    group_by(Protein_Accession) %>%
    summarise(
      InterPro_Accession = {
        x <- unique(InterPro_Accession)
        if (all(x == "-")) NA_character_ else paste(x[x != "-"], collapse = "; ")
      },
      InterPro_Description = {
        x <- unique(InterPro_Description)
        if (all(x == "-")) NA_character_ else paste(x[x != "-"], collapse = "; ")
      },
      GO_Terms = {
        x <- unique(GO_Terms)
        if (all(x == "-")) NA_character_ else paste(x[x != "-"], collapse = "; ")
      },
      .groups = "drop"
    )
  
  name <- tools::file_path_sans_ext(basename(f))
  summary_list[[name]] <- summary_df
  
  # Optional: save CSV here if you want
}

# After the loop, you can combine or do other things with summary_list


```

```{r}
combined_df <- bind_rows(
  lapply(names(summary_list), function(n) {
    summary_list[[n]] %>% mutate(source_file = n)
  })
)

```

```{r}
combined_df
```

```{r}
write.csv(combined_df, "interpro_summary1.csv", row.names = FALSE)

```

let's add annotation categories to my heatmap

```{r}
logfc_matrix
```


```{r}
#without annotation categories
pheatmap(logfc_matrix,
         cluster_rows = TRUE,
         cluster_cols = TRUE,
         scale = "none",  # or "row" for z-scores per gene
         color = colorRampPalette(c("blue", "white", "orange"))(100),
         fontsize_row = 8,  # Adjust font size for row labels
         angle_col = 0)    # Rotate column labels to 0° (horizontal)
```

```{r}
#this is the list of row names in order
#now need to make list of each category name in order 

#"21469: Cytochrome P450 monooxygenase", "27306: dormancy/auxin-related", "31667: GMC oxidoreductase", "38503: CRK with Gnk2 domain", "30375: R13L4/SHOC-2-like", "29048: DUF829", "51370: GMC oxidoreductase", "27294: FAD/NAD(P)-binding monooxygenase", "23020: unknown", "37466: Myc-like bHLH transcription factor", "13638: DCN1-like with PONY-binding domain", "42174: siRNA processing", "19388: UDP-glycosyltransferase", "51302: UspA", "06003: MKZ1-like", "12596: myo-insitol-1-phosphate synthase", "16206: Fe(2+)/2OG dioxygenase", "47136: XTH", "12910: pectinesterase inhibitor", "04115: CRISP with Gnk2 domain", "04789: sulfotransferase", "31428: TauD/TfdA-like", "11752: MFS sugar transporter", "08506: JINGUBANG-like with WD40-repeat", "39647: CRK with Gnk2 domain", "48863: MFS myo-inositol transporter", "40178: BURP domain", "38415: UspA", "40947: GH16"

```


start with making df of annotation categories:
```{r}
gene_categories <- data.frame(
  Category = c("metabolism", "development", "metabolism", "cell signaling", "defense", "unknown", "metabolism", "metabolism", "unknown", "gene regulation", "development", "gene regulation", "metabolism", "stress response", "development", "cell signaling", "metabolism", "cell wall mod", "cell wall mod", "defense", "metabolism", "metabolism", "sugar transport", "cell signaling", "cell signaling", "sugar transport", "cell wall mod", "stress response", "cell wall mod")
)
rownames(gene_categories) <- rownames(logfc_matrix)  # must match order in matrix

colnames(gene_categories) <- " "  # remove the label entirely


```


```{r}
annotation_colors <- list(
  " " = c(
    "metabolism" = "#1f77b4",      # blue
    "stress response" = "#ff7f0e",      # orange
    "sugar transport" = "#2ca02c",         # green
    "cell wall mod" = "#d62728",    # red
    "unknown" = "#7f7f7f",         # gray
    "gene regulation" = "#9467bd",
    "cell signaling" = "#8c564b", 
    "defense" = "#e377c2", 
    "development" = "#17becf" 
  )
)

```


```{r}
pheatmap(logfc_matrix,
         cluster_rows = TRUE,
         cluster_cols = TRUE,
         scale = "none",
         color = colorRampPalette(c("blue", "white", "orange"))(100),
         fontsize_row = 8,
         angle_col = 0,
         annotation_row = gene_categories,
         annotation_colors = annotation_colors)

```

```{r}
png("goodheatmap.png", width = 1200, height = 1600, res = 150)
pheatmap(logfc_matrix,
         cluster_rows = TRUE,
         cluster_cols = TRUE,
         scale = "none",
         color = colorRampPalette(c("blue", "white", "orange"))(100),
         fontsize_row = 8,
         angle_col = 0,
         annotation_row = gene_categories,
         annotation_colors = annotation_colors)
dev.off()

```





