---
title: "Demorie RNA Analysis Workflow"
author: "Britt"
date: "2024-06-14"
output: html_document
editor_options: 
  markdown: 
    wrap: sentence
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

**Basic steps for RNAseq Data Processing:**

1.  Check FASTQ quality.
2.  Trim reads to keep high-quality reads.
3.  Split reads based on barcodes.
4.  Map to reference genome.

(can also map to transcriptome)

**Basic steps for DGE Analysis:**

1.  Load the RNAseq counts.
2.  Normalize the counts.
3.  QC the counts.
4.  Create a data frame that describes the experiment.
5.  Determine the dispersion parameter (how over-dispersed is the data?)
6.  Fit a statistical model for gene expression as a function of experimental parameters.
7.  Test the significance of experimental parameters for explaining gene expression.
8.  Examine result.

## Processing RNAseq Data

Programs needed: FASTQC, Trimmomatic, auto_barcode, STAR

### Index the Reference Genome

```{bash, eval = FALSE}
#download reference genome from personal computer
#this code goes in the personal computer terminal
$ scp -P 22345 /c/Users/britt/Documents/Trucit1_1_AssemblyScaffolds.fasta.gz bklong@bogarprecision.plb.ucdavis.edu:~/DemRNAanalysis/TRCIindex
```

```{bash, eval = FALSE}
#make sure reference genome is unzipped
gunzip -k Psme.1_0.fa.gz
gunzip -k Trucit1_1_AssemblyScaffolds.fasta.gz
```

```{bash, eval = FALSE}
#or use hisat?
hisat2 -x /data/Genomes/HISAT2_index/Psme_1_0_index -1 /data/Test/DFD_001_T14_R1.fastq.gz -2 /data/Test/DFD_001_T14_R2.fastq.gz -S output.sam --summary-file summary.txt --new-summary --mm

# Run STAR genome index generation
~/programs/STAR-2.7.11b/bin/Linux_x86_64_static/STAR --runMode genomeGenerate \
           --genomeDir ~/DemRNAanalysis/PSMEindex \
           --genomeFastaFiles ~/DemRNAanalysis/genome/Psme.1_0.fa \
           --runThreadN 4 \
           --limitGenomeGenerateRAM 6425695398837
```

```{bash, eval = FALSE}
~/programs/STAR-2.7.11b/bin/Linux_x86_64_static/STAR --runMode genomeGenerate \
           --genomeDir ~/DemRNAanalysis/TRCIindex \
           --genomeFastaFiles ~/DemRNAanalysis/TRCIindex/Trucit1_1_AssemblyScaffolds.fasta \
           --runThreadN 4 \
           --limitGenomeGenerateRAM 6425695398837
           
#need to modify overhang parameter (max read length - 1)
#need annotation file for TRCI
STAR \
    --runThreadN 7 \
    --runMode genomeGenerate \
    --genomeDir TRCIindex \
    --genomeSAindexNbases 12 \
    --sjdbOverhang 93 \ 
    --sjdbGTFfile Brapa_gene_v1.5.gff \
    --sjdbGTFfeatureExon CDS \
    --sjdbGTFtagExonParentTranscript Parent \
    --genomeFastaFiles Trucit1_1_AssemblyScaffolds.fasta
```

### Check Quality of Reads

```{bash, eval = FALSE}
fastqc
#then load file of interest into user interface
```

### Trim Low Quality Reads if Necessary

```{bash, eval = FALSE}
#this is an example
trimmomatic SE GH.lane67.fastq.gz GH.lane67.trimmed.fastq SLIDINGWINDOW:4:20 MINLEN:50
trimmomatic SE GH.lane67.fastq.gz GH.lane67.trimmed.fastq SLIDINGWINDOW:4:20 MINLEN:50
trimmomatic SE -threads 1 GH.lane67.fastq.gz GH.lane67.trimmed.fastq SLIDINGWINDOW:4:20 MINLEN:50
trimmomatic SE -threads 1 GH.lane67.fastq.gz GH.lane67.trimmed.fastq SLIDINGWINDOW:4:20 MINLEN:50
```

If you trim reads, make sure you put them through FASTQC again to see if it worked.

### Splitting by Barcodes

```{bash, eval = FALSE}
#use auto_barcode to split reads
```

### Check Quality of Libraries

```{bash, eval = FALSE}
fastqc --threads 7 -o ../../output/fastqc_out split_fq/*.fq

multiqc fastqc_out
```

### Map Reads to Reference

```{bash, eval = FALSE}
#modify this to be a for loop for all fastq files
time STAR --runThreadN 7 \
--genomeDir input/Brapa_reference/Brapa_STAR_index \
--readFilesIn input/Brapa_fastq/split_fq/IMB211_DP_1_SILIQUE.fq \
--outFileNamePrefix output/STARout/IMB211_DP_1_SILIQUE.fq_ \
--outSAMtype BAM SortedByCoordinate \
--outBAMsortingThreadN 2 \
--alignIntronMax 5000
```

## Differential Gene Expression

Programs needed: HTseq-count

### Get Read Counts

```{bash, eval = FALSE}
#example
htseq-count -s no -t CDS -i Parent \
../../assignment-09-jnmaloof/output/STAR_out-IMB211_INTERNODE_A03/IMB211_INTERNODE_Aligned_A03.bam \
../../assignment-09-jnmaloof/output/STAR_out-R500_INTERNODE_A03/R500_INTERNODE_Aligned_A03.bam \
../../assignment-09-jnmaloof/input/Brapa_reference/Brapa_gene_v1.5.gff > A03_counts.tsv
```

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

### DGE With Counts Data

#### (start here after bioinformatics core returns counts data)

```{bash, eval = FALSE}
#take a look at counts data
head A03_counts.tsv #replace this file with what you get back from core
```

```{r}
library(readr)
counts.data <- read_tsv("../input/gh_internode_counts2.tsv") #assign counts data to an object
```

"You may have noticed that the first gene_id is labelled \"\*\". These are the reads that did not map to a gene. Let\'s remove this row from the data. Also let\'s replace any \"NA\" records with \"0\" because that is what NA means in this case."

```{r}
#change NAs to 0s

library(tidyverse)

counts.data <- counts.data %>% filter(gene_id!="*")
counts.data[is.na(counts.data)] <- 0

head(counts.data)
```

"The column names are too long. Use the `str_remove()` command to remove the ".1_matched.merged.fq.bam" suffix from each column name. Although it doesn't matter in this case, surrounding the "pattern" inside of the function `fixed()` would be a good idea, because "." is a wildcard character."

```{r}
#shorten column names
library(stringr)

colnames(counts.data) <- str_remove(colnames(counts.data), fixed(".1_matched.merged.fq.bam"))

colnames(counts.data)
```

```{r}
#retain only the reads that were over 10 count in at least three samples
counts.data <- counts.data[rowSums(counts.data[,-1] > 10) >= 3,]
```

"We expect that read counts, especially from biological replicates, will be highly correlated. Check to see if this is the case using the `pairs()` function and the `cor()` function. Pairs should be applied to the (potentially log transformed) counts.data, NOT the correlation matrix."

```{r}
#the below code needs to be fixed
counts_subset <- counts.data[, -which(names(counts.data) == "gene_id")]
counts_subset <- counts_subset[1:1000, ]
correlation <- cor(counts_subset)
pairs(correlation)

transformed_counts <- log2(counts_subset + 1) 
transformed_correlation <- cor(transformed_counts)
pairs(transformed_correlation)
```

"Once you have a correlation table, use the code provided to visualize it. Then, comment on the results from the pairs and correlation heatmap plots. Are the replicates more similar to each other than they are to other samples? Do you think there are any mistakes in the sample treatment labels?"

```{r}
rownames(cor.table) <- str_remove(rownames(cor.table), "_INTERNODE.*") #shorter names for better plotting
colnames(cor.table) <- str_remove(colnames(cor.table), "_INTERNODE.*")

cor.table %>% gplots::heatmap.2(dendrogram="row", trace = "none", col=viridis::viridis(25, begin=.25), margins=c(7,8))
```

Data Normalization:
