---
title: "Demorie RNA Analysis Workflow"
author: "Britt"
date: "2024-06-14"
output: html_document
editor_options: 
  markdown: 
    wrap: sentence
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

**Basic steps for RNAseq Data Processing:**

1.  Check FASTQ quality.
2.  Trim reads to keep high-quality reads.
3.  Split reads based on barcodes.
4.  Map to reference genome.

(can also map to transcriptome)

**Basic steps for DGE Analysis:**

1.  Load the RNAseq counts.
2.  Normalize the counts.
3.  QC the counts.
4.  Create a data frame that describes the experiment.
5.  Determine the dispersion parameter (how over-dispersed is the data?) 6. Fit a statistical model for gene expression as a function of experimental parameters.
6.  Test the significance of experimental parameters for explaining gene expression.
7.  Examine result.

## Processing RNAseq Data

Programs needed: FASTQC, Trimmomatic, auto_barcode, 

### Index the Reference Genome


