---
title: "Demorie RNA Analysis Workflow"
author: "Britt"
date: "2024-06-14"
output: html_document
editor_options: 
  markdown: 
    wrap: sentence
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

**Basic steps for RNAseq Data Processing:**

1.  Check FASTQ quality.
2.  Trim reads to keep high-quality reads.
3.  Split reads based on barcodes.
4.  Map to reference genome.

(can also map to transcriptome)

**Basic steps for DGE Analysis:**

1.  Load the RNAseq counts.
2.  Normalize the counts.
3.  QC the counts.
4.  Create a data frame that describes the experiment.
5.  Determine the dispersion parameter (how over-dispersed is the data?)
6.  Fit a statistical model for gene expression as a function of experimental parameters.
7.  Test the significance of experimental parameters for explaining gene expression.
8.  Examine result.

## Processing RNAseq Data

Programs needed: FASTQC, Trimmomatic, auto_barcode, STAR

### Index the Reference Genome

```{bash, eval = FALSE}
#download reference genome from personal computer
#this code goes in the personal computer terminal
$ scp -P 22345 /c/Users/britt/Documents/Trucit1_1_AssemblyScaffolds.fasta.gz bklong@bogarprecision.plb.ucdavis.edu:~/DemRNAanalysis/TRCIindex
```

```{bash, eval = FALSE}
#make sure reference genome is unzipped
gunzip -k Psme.1_0.fa.gz
gunzip -k Trucit1_1_AssemblyScaffolds.fasta.gz
```

```{bash, eval = FALSE}
#or use hisat?
hisat2 -x /data/Genomes/HISAT2_index/Psme_1_0_index -1 /data/Test/DFD_001_T14_R1.fastq.gz -2 /data/Test/DFD_001_T14_R2.fastq.gz -S output.sam --summary-file summary.txt --new-summary --mm

# Run STAR genome index generation
~/programs/STAR-2.7.11b/bin/Linux_x86_64_static/STAR --runMode genomeGenerate \
           --genomeDir ~/DemRNAanalysis/PSMEindex \
           --genomeFastaFiles ~/DemRNAanalysis/genome/Psme.1_0.fa \
           --runThreadN 4 \
           --limitGenomeGenerateRAM 6425695398837
```

```{bash, eval = FALSE}
~/programs/STAR-2.7.11b/bin/Linux_x86_64_static/STAR --runMode genomeGenerate \
           --genomeDir ~/DemRNAanalysis/TRCIindex \
           --genomeFastaFiles ~/DemRNAanalysis/TRCIindex/Trucit1_1_AssemblyScaffolds.fasta \
           --runThreadN 4 \
           --limitGenomeGenerateRAM 6425695398837
           
#need to modify overhang parameter (max read length - 1)
#need annotation file for TRCI
STAR \
    --runThreadN 7 \
    --runMode genomeGenerate \
    --genomeDir TRCIindex \
    --genomeSAindexNbases 12 \
    --sjdbOverhang 93 \ 
    --sjdbGTFfile Brapa_gene_v1.5.gff \
    --sjdbGTFfeatureExon CDS \
    --sjdbGTFtagExonParentTranscript Parent \
    --genomeFastaFiles Trucit1_1_AssemblyScaffolds.fasta
```

### Check Quality of Reads

```{bash, eval = FALSE}
fastqc
#then load file of interest into user interface
```

### Trim Low Quality Reads if Necessary

```{bash, eval = FALSE}
#this is an example
trimmomatic SE GH.lane67.fastq.gz GH.lane67.trimmed.fastq SLIDINGWINDOW:4:20 MINLEN:50
trimmomatic SE GH.lane67.fastq.gz GH.lane67.trimmed.fastq SLIDINGWINDOW:4:20 MINLEN:50
trimmomatic SE -threads 1 GH.lane67.fastq.gz GH.lane67.trimmed.fastq SLIDINGWINDOW:4:20 MINLEN:50
trimmomatic SE -threads 1 GH.lane67.fastq.gz GH.lane67.trimmed.fastq SLIDINGWINDOW:4:20 MINLEN:50
```

If you trim reads, make sure you put them through FASTQC again to see if it worked.

### Splitting by Barcodes

```{bash, eval = FALSE}
#use auto_barcode to split reads
```

### Check Quality of Libraries

```{bash, eval = FALSE}
fastqc --threads 7 -o ../../output/fastqc_out split_fq/*.fq

multiqc fastqc_out
```

### Map Reads to Reference

```{bash, eval = FALSE}
#modify this to be a for loop for all fastq files
time STAR --runThreadN 7 \
--genomeDir input/Brapa_reference/Brapa_STAR_index \
--readFilesIn input/Brapa_fastq/split_fq/IMB211_DP_1_SILIQUE.fq \
--outFileNamePrefix output/STARout/IMB211_DP_1_SILIQUE.fq_ \
--outSAMtype BAM SortedByCoordinate \
--outBAMsortingThreadN 2 \
--alignIntronMax 5000
```

## Differential Gene Expression

Programs needed: HTseq-count

### Get Read Counts

```{bash, eval = FALSE}
#example
htseq-count -s no -t CDS -i Parent \
../../assignment-09-jnmaloof/output/STAR_out-IMB211_INTERNODE_A03/IMB211_INTERNODE_Aligned_A03.bam \
../../assignment-09-jnmaloof/output/STAR_out-R500_INTERNODE_A03/R500_INTERNODE_Aligned_A03.bam \
../../assignment-09-jnmaloof/input/Brapa_reference/Brapa_gene_v1.5.gff > A03_counts.tsv
```
