---
title: "Siyu index genome"
author: "Britt"
date: "2024-07-24"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Counts reads in our data to get an idea of how big the files are:

```{bash, eval = FALSE}
#this is imperfect and needs to be modified to tell us how many reads per fastq file
for file in *.fastq.gz; do
    echo "Processing file: $file"
    # Command to count reads in the fastq file
    read_count=$(zcat "$file" | echo $(( $(wc -l) / 4 )))
    echo "Reads in $file: $read_count"
done
```

Download the TRCI transcriptome on personal computer and move it to server:

```{bash, eval = FALSE}
#this code goes in the personal computer terminal
scp -P 22345 /c/Users/britt/Documents/Trucit1_1_EST_20220613_cluster_consensi.fasta.gz bklong@bogarprecision.plb.ucdavis.edu:~/DemRNAanalysis/TRCItranscriptome
```

Using Kallisto to index genome:

Required argument:
-i, --index=STRING
Add a file name for the kallisto index to be constructed.

```{bash, eval = FALSE}
#cd to location of fasta file
kallisto index -i TRCI.idx Trucit1_1_EST_20220613_cluster_consensi.fasta.gz
```

Using Kallisto to produce counts data:

Required arguments:
-i, --index=STRING            
Filename for the kallisto index to be used for quantification.

-o, --output-dir=STRING       
Directory to write output to.

kallisto can process either single-end or paired-end reads. The default running mode is paired-end and requires an even number of FASTQ files represented as pairs. Important note: only supply one sample at a time to kallisto. The multiple FASTQ (pair) option is for users who have samples that span multiple FASTQ files.

running this on just one sample for now using screen to make sure it keeps running:

```{bash, eval = FALSE}
screen -S kallisto #makes new screen
#Ctrl+A, then D disconnects
screen -r kallisto #reconnects
#will run in background even if you close the terminal
```


```{bash, eval = FALSE}
kallisto quant -i TRCI.idx -o DFD_019_counts DFD_019_T9_R1.fastq.gz DFD_019_T9_R2.fastq.gz
```

took about an hour and a half to run above code, but seemingly worked!

let's write a for loop to automate this for us:

```{bash, eval = FALSE}
#!/bin/bash

# Define the directory containing the FASTQ files
FASTQ_DIR="path/to/your/fastq/files"

# Define the Kallisto index file
KALLISTO_INDEX="TRCI.idx"

# Loop over each R1 FASTQ file in the directory
for R1_FILE in "$FASTQ_DIR"/*_R1.fastq.gz; do
  # Generate the corresponding R2 file name by replacing _R1 with _R2
  R2_FILE="${R1_FILE/_R1/_R2}"
  
  # Extract the base name of the file to create the output directory name
  BASE_NAME=$(basename "$R1_FILE" _R1.fastq.gz)
  
  # Define the output directory for Kallisto
  OUTPUT_DIR="${FASTQ_DIR}/${BASE_NAME}_counts"
  
  # Run Kallisto quant
  kallisto quant -i "$KALLISTO_INDEX" -o "$OUTPUT_DIR" "$R1_FILE" "$R2_FILE"
done
```

without the comments, broken up into two steps:

Here are the fungal samples - 19,26,27,30,45,52,62,87

```{bash, eval = FALSE}
FASTQ_DIR="/fungal_data"

KALLISTO_INDEX="TRCI.idx"
```

```{bash}
for R1_FILE in "$FASTQ_DIR"/*_R1.fastq.gz; do
  R2_FILE="${R1_FILE/_R1/_R2}"
  
  BASE_NAME=$(basename "$R1_FILE" _R1.fastq.gz)
  
  OUTPUT_DIR="${FASTQ_DIR}/${BASE_NAME}_counts"
 
  kallisto quant -i "$KALLISTO_INDEX" -o "$OUTPUT_DIR" "$R1_FILE" "$R2_FILE"
done
```

----------------
My first attempt at mapping reads from sample 19 was successful!

I will not move the files to my personal computer to analyze them further with R.

Move files to personal computer to look at/analyze:

```{bash, eval = FALSE}
#this code goes in the personal computer terminal
scp -P 22345 bklong@bogarprecision.plb.ucdavis.edu:/data/fungal_data/DFD_019_counts/abundance.tsv /c/Users/britt/Documents/R/TRCI_counts
```





 