---
title: "Siyu index genome"
author: "Britt"
date: "2024-07-24"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Counts reads in our data to get an idea of how big the files are:

```{bash, eval = FALSE}
#this is imperfect and needs to be modified to tell us how many reads per fastq file
for file in *.fastq.gz; do
    echo "Processing file: $file"
    # Command to count reads in the fastq file
    read_count=$(zcat "$file" | echo $(( $(wc -l) / 4 )))
    echo "Reads in $file: $read_count"
done
```

Download the TRCI transcriptome on personal computer and move it to server:

```{bash, eval = FALSE}
#this code goes in the personal computer terminal
scp -P 22345 /c/Users/britt/Documents/Trucit1_1_EST_20220613_cluster_consensi.fasta.gz bklong@bogarprecision.plb.ucdavis.edu:~/DemRNAanalysis/TRCItranscriptome
```

Using Kallisto to index genome:

Required argument:
-i, --index=STRING
Add a file name for the kallisto index to be constructed.

```{bash, eval = FALSE}
#cd to location of fasta file
kallisto index -i TRCI.idx Trucit1_1_EST_20220613_cluster_consensi.fasta.gz
```

Using Kallisto to produce counts data:

Required arguments:
-i, --index=STRING            
Filename for the kallisto index to be used for quantification.

-o, --output-dir=STRING       
Directory to write output to.

```{bash, eval = FALSE}
kallisto quant -i TRCI.idx -o TRCI_counts pairA_1.fastq pairA_2.fastq pairB_1.fastq pairB_2.fastq
```









 