---
title: "STAR"
author: "Britt"
date: "2024-05-07"
output: html_document
---

#starting to code

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

making a genome index:

```{bash}
# Define the path to the STAR executable
STAR_PATH=/path/to/STAR

# Define the path to the directory where you want to save the genome index
GENOME_INDEX_DIR=/path/to/genome_index

# Define the path to the reference genome FASTA file
GENOME_FASTA=/path/to/reference_genome.fasta

# Run STAR genome index generation
$STAR_PATH --runMode genomeGenerate \
           --genomeDir $GENOME_INDEX_DIR \
           --genomeFastaFiles $GENOME_FASTA \
           --runThreadN 4

```

running STAR:

```{bash}
# Define the path to the STAR executable
STAR_PATH=/path/to/STAR

# Define the path to the reference genome index
GENOME_INDEX=/DemRNAanalysis/genome

# Define the path to the directory containing the input FASTQ files
INPUT_DIR=/DemRNAanalysis/seqs

# Define the path to the output directory where STAR output will be saved
OUTPUT_DIR=/DemRNAanalysis/STARoutput

# Run STAR alignment
$STAR_PATH --genomeDir $GENOME_INDEX \
           --readFilesIn $INPUT_DIR/reads1.fastq $INPUT_DIR/reads2.fastq \
           --outFileNamePrefix $OUTPUT_DIR/output_prefix \
           --runThreadN 4 \
           --outSAMtype BAM SortedByCoordinate \
           --quantMode GeneCounts

# Explanation of options:
# --genomeDir: Path to the directory containing the genome index files
# --readFilesIn: Path to the input FASTQ files (reads1 and reads2)
# --outFileNamePrefix: Prefix for the output file names
# --runThreadN: Number of threads to use for alignment
# --outSAMtype: Output format (BAM SortedByCoordinate)
# --quantMode: Output gene counts

```
