---
title: "TO.16S.analysis"
author: "Britt"
date: "2025-01-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This is the document I will use to analyze 16S TO data in Christy's microbial metagenomics class.

```{r}
install.packages("fastqcr")
```
```{r}
library(fastqcr)
```

```{r, eval = FALSE}
#run fastqc
fastqc(fq.dir = "16S/TO16S_filtered", qc.dir = "16SQC",threads = 4)
#summarize
qc <- qc_aggregate("16SQC")
#view summary
qc
```
fq.dir = location you have your fastq files

qc.dir = location you want the output

Oops, looks like this doesn't work on my windows computer. That's okay, all fastqc does is visualize the quality of our data. Looks like I will skip this step and go straight into quality control! Chirsty says that generally we don't even bother doing fastqc on 16S data.

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
```

```{r}
BiocManager::install("dada2")
```

```{r}
library("dada2")
```

```{r}
path <- "16S/TO16S_filtered" # CHANGE ME to the directory containing the fastq files after unzipping.
list.files(path)
```

```{r}
fnFs <- sort(list.files(path, pattern="_R1.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_R2.fastq.gz", full.names = TRUE))

# Extract sample names from filenames
sample.names <- sapply(basename(fnFs), function(x) {
  parts <- strsplit(x, "_")[[1]]
  paste(parts[3], parts[4], parts[6], parts[7], sep = "_")
})

# Remove the names from the vector
names(sample.names) <- NULL

# Print the sample names
print(sample.names)
```

```{r}
plotQualityProfile(fnFs[1:2])
```

```{r}
plotQualityProfile(fnRs[1:2])
```

```{r}
# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names
```

Before you run this line, change the values to where you say is a good cutoff for your data: 

```{r}
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(300,300),maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,compress=TRUE, multithread=TRUE, trimLeft = c(19, 20))
head(out)
```

#(for above): I left multithread=TRUE in, but for windows, can remove it. It threw an error that said multithreading was disabled. So, I am assuming it is working now.

If the above code ever finishes, I can save my progress by outputting it as an RDS that I can load if need be:

```{r}
saveRDS(out, "out.rds")
saveRDS(filtFs, "filtFs.rds")
saveRDS(filtRs, "filtRs.rds")
```

do not run this next chunk. It's another FASTQC that we don't necessarily need. Skip.

```{r, eval = FALSE}
fastqc(fq.dir = "DATA/filtered/", qc.dir = "FASTQC_2",threads = 4) 
qc2 <- qc_aggregate("FASTQC_2") 
qc2

plotQualityProfile(filtFs[1:2])
plotQualityProfile(filtRs[1:2])
```

```{r}
errF <- learnErrors(filtFs, multithread=TRUE) # creates an object that contains the errors for the forward sequences
errR <- learnErrors(filtRs, multithread=TRUE)# creates an object that contains the errors for the reverse sequences
```

```{r}
saveRDS(errF, "errF.rds")
saveRDS(errR, "errR.rds")
```


```{r}
plotErrors(errF, nominalQ=TRUE) # this plots the forward error rates
```

```{r}
plotErrors(errR, nominalQ=TRUE) # this plots the reverse error rates
```

There are 420 sequence files (both F and R), so therefore there should be 210 samples. That means I expect the output of both the above and below code sections to result in 210 samples each. 

dereplication

```{r}
derepFs <- derepFastq(filtFs, verbose=TRUE)
derepRs <- derepFastq(filtRs, verbose=TRUE)
# Name the derep-class objects by the sample names
names(derepFs) <- sample.names
names(derepRs) <- sample.names
```

```{r}
#saved to computer but did not push. pushing yeeted R
saveRDS(derepFs, "derepFs.rds")
saveRDS(derepRs, "derepRs.rds")
```

```{r}
derepFs <- readRDS("derepFs.rds")
derepRs <- readRDS("derepRs.rds")
```

```{r}
errF <- readRDS("errF.rds")
errR <- readRDS("errR.rds")
```


```{r}
dadaFs <- dada(derepFs, err = errF, verbose = TRUE, multithread = TRUE)
```
It finally worked! It took about 1.5 hours to run above code. Now let's try the next chunk...

```{r}
dadaRs <- dada(derepRs, err=errR, verbose = TRUE, multithread=TRUE)
```

It worked! This one was much faster, maybe 1 hr.

```{r}
saveRDS(dadaFs, "dadaFs.rds")
```

```{r}
saveRDS(dadaRs, "dadaRs.rds")
```

```{r}
dadaFs <- readRDS("dadaFs.rds")
dadaRs <- readRDS("dadaRs.rds")
```

----------------------------------------------------------
OR TO POOL:

pseudo pool

```{r, eval = FALSE}
dadaFs.pool <- dada(derepFs, err = errF, verbose = TRUE, multithread = TRUE, pool = "pseudo")
```

```{r, eval = FALSE}
dadaRs.pool <- dada(derepRs, err = errF, verbose = TRUE, multithread = TRUE, pool = "pseudo")
```
------------------------------------------------------------

start here after saving dada RDS:

```{r}
mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE)
head(mergers[[1]])
```

```{r}
mergers <- readRDS("mergers.rds")
```

```{r}
seqtab <- makeSequenceTable(mergers) # make sequence table
dim(seqtab) # tell me the dimensions of the sequence table.
```

```{r}
table(nchar(getSequences(seqtab)))
```

Optional Step: Edit 250:256 below to indicate the range of sequence lengths that you would like to keep. Otherwise, delete the chunk or comment out the lines. (I will skip this step due to the new type of primer we used).

```{r, eval = FALSE}
seqtab2 <- seqtab[,nchar(colnames(seqtab)) %in% 250:256] # remove sequences shorter than 250 (as written) and longer than 256 (as written)
sum(seqtab2)/sum(seqtab)
seqtab <- seqtab2 # rename the object you made above to seqtab so you can continue this tutorial without complications. For your own work, I would suggest leaving it named seqtab2 and using "seqtab2" instead of "seqtab" for all steps below. This will help you remember what you were did.
```

remove chimeras:

```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
```


```{r}
saveRDS(seqtab.nochim, "seqtab.nochim.rds")
```


let's take a look!

```{r}
dim(seqtab.nochim)
```

```{r}
sum(seqtab.nochim)/sum(seqtab)
```

sanity check: this tracks how our data changed over quality control to double check that everything seems reasonable. You will expect to see a big drop in the filtered column, but across the row for a sample, shouldn't change much except in filtered.

```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
track
```
The above output looks weird...lots of losses, but we will push forward for now.

```{r}
saveRDS(track, "track.rds")
```


```{r}
seqtab.nochim <- readRDS("seqtab.nochim.rds")
```

```{r}
library(dada2)
```


Assign Taxonomy

Download SILVA database first. Change path to windows path by finding file and copying path, then pasting. Change to forward slashed and remove C:.

```{r}
taxa <- assignTaxonomy(seqtab.nochim, "/Users/britt/Downloads/silva_nr99_v138.2_toSpecies_trainset.fa.gz")
taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
```
It took 5 hours to run the above chunk.

```{r}
saveRDS(taxa, "taxa.rds")
```


```{r}
install.packages("phyloseq")
```


```{r}

library(Biostrings); packageVersion("Biostrings")
library(ggplot2); packageVersion("ggplot2")
library("phyloseq"); packageVersion("phyloseq")
theme_set(theme_bw())

library("microbiome")
```

```{r}
library("phyloseq")
```

looks like I need to add my metadata below before running
```{r}
metadata <- read.csv("16S/16S.metadata.csv", header=TRUE, row.names=1)

ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(metadata), 
               tax_table(taxa))

dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))
ps
```

```{r}
saveRDS(ps, "ps__.rds")
```

Use ps__.rds if you need to reload it.

Removing Unwanted Data

Edit the code below to remove the taxa that are not targeted by your primers. By the end, you want your phyloseq object to be named ps_final.

The below code removes all chloroplasts, all mitochondira, and anything that is NOT a bacteria. 

```{r remove unwanted, results="hide", fig.show='hide', warning=FALSE, message=FALSE}
  ps1 <- subset_taxa(ps, (Order!="Chloroplast") | is.na(Class))
  ps2 <- subset_taxa(ps1, (Family!="Mitochondria") | is.na(Family))
ps_bacteria <- subset_taxa(ps2, Kingdom =="Bacteria")
ps_final <-ps_bacteria
```

```{r}
saveRDS(ps_final, "ps_final.rds")
```


Analyzing the Data

rarefaction curve:

if flat, we sampled the diversity of our community!
if still sloped, we didn't sample the diversity of our community! 
```{r}
#install.packages(vegan)
library(vegan); packageVersion("vegan")
table <- otu_table(ps_final)
class(table) <- "matrix"
rarecurve(table, step=1000, cex=0.5)
```

to describe, use full dataset
to compare communities between treatments, rarify (subsample)

bar graph / community composition:

```{r}
ps_final <- prune_samples(sample_sums(ps_final) > 0, ps_final)

ps_phylum <- tax_glom(ps_final, taxrank="Phylum")
top10 <- names(sort(taxa_sums(ps_phylum), decreasing=TRUE))[1:10]
ps.top10 <- transform_sample_counts(ps_phylum, function(OTU) OTU/sum(OTU))
ps.top10 <- prune_taxa(top10, ps.top10)
plot_bar(ps.top10, fill="Phylum") 
```




look at community diversity:

richness = # of species
evenness = how many individuals in each of the species (more diverse if each species has a more even amount of individuals)
```{r}
plot_richness(ps_final, x=all_quantile, measures=c("Shannon", "Simpson"), color="site", shape="site")
```

Note: Unifrac is helpful to look at diversity while accounting for OTUs being more similar to each other phylogenetically (but requires a phylogenetic tree)...we will skip this for this assignment

Comparing Communities

```{r}
ord.nmds.bray <- ordinate(ps_final, method="NMDS", distance="bray")
ord.nmds.bray
plot_ordination(ps_relabund, ord.nmds.bray, title="Bray NMDS",   color=as.factor("site", shape="site"))
```

```{r}
ps_phylum <- tax_glom(ps_bacteria, taxrank = "Phylum")
ps_class <- tax_glom(ps_bacteria, taxrank = "Class")
ps_family <- tax_glom(ps_bacteria, taxrank = "Family")
ps_genus <- tax_glom(ps_bacteria, taxrank = "Genus")
```




